<!DOCTYPE html>

<html lang="en" data-content_root="../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>preprocessing_pipeline.download &#8212; scDoRI  documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css?v=27fed22d" />
    <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for preprocessing_pipeline.download</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="download_file">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.download.download_file">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">download_file</span><span class="p">(</span><span class="n">url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download a file from a given URL and save it locally using the `wget` command.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    url : str</span>
<span class="sd">        The URL to download from.</span>
<span class="sd">    out_path : pathlib.Path</span>
<span class="sd">        The local file path where the downloaded file will be saved.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        If the file already exists, no download is performed. Otherwise, runs</span>
<span class="sd">        a shell command with `wget` to fetch the file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">out_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File already exists: </span><span class="si">{</span><span class="n">out_path</span><span class="si">}</span><span class="s2">. Skipping download.&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;wget -O </span><span class="si">{</span><span class="n">out_path</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Downloading </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">out_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="unzip_gz">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.download.unzip_gz">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">unzip_gz</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">remove_input</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Decompress a GZIP file (i.e., `.gz`) in-place using `gzip -d`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file_path : pathlib.Path</span>
<span class="sd">        Path to the `.gz` file to be decompressed.</span>
<span class="sd">    remove_input : bool, optional</span>
<span class="sd">        If True, delete the original `.gz` file after successful decompression.</span>
<span class="sd">        Default is False.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        Decompresses the file in-place, leaving a new file without the `.gz` extension.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;gzip -d </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Decompressing </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">remove_input</span><span class="p">:</span>
        <span class="n">gz_file</span> <span class="o">=</span> <span class="n">file_path</span>
        <span class="k">if</span> <span class="n">gz_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">gz_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span></div>


<div class="viewcode-block" id="resolve_genome_urls">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.download.resolve_genome_urls">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">resolve_genome_urls</span><span class="p">(</span>
    <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gtf_url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">chrom_sizes_url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">fasta_url</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Resolve final URLs for GTF, chromosome sizes, and FASTA files based on species and assembly.</span>

<span class="sd">    1. If user provides URLs (gtf_url, chrom_sizes_url, fasta_url), use them directly.</span>
<span class="sd">    2. If not, attempt to use known defaults for recognized combos (e.g. mouse/mm10, human/hg38).</span>
<span class="sd">    3. If the combo is unrecognized and the user hasn&#39;t provided all URLs, raise an error.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    species : str</span>
<span class="sd">        Species name (e.g. &quot;mouse&quot;, &quot;human&quot;).</span>
<span class="sd">    assembly : str</span>
<span class="sd">        Genome assembly (e.g. &quot;mm10&quot;, &quot;hg38&quot;).</span>
<span class="sd">    gtf_url : str or None</span>
<span class="sd">        A user-provided URL for the GTF file, or None to try defaults.</span>
<span class="sd">    chrom_sizes_url : str or None</span>
<span class="sd">        A user-provided URL for the chromosome sizes file, or None to try defaults.</span>
<span class="sd">    fasta_url : str or None</span>
<span class="sd">        A user-provided URL for the FASTA file, or None to try defaults.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    (final_gtf_url, final_chrom_sizes_url, final_fasta_url) : tuple of str</span>
<span class="sd">        The resolved URLs for GTF, chromosome sizes, and FASTA. These may come from</span>
<span class="sd">        user-provided values or known defaults if recognized. If no defaults</span>
<span class="sd">        exist and user inputs are missing, raises ValueError.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">final_gtf_url</span> <span class="o">=</span> <span class="n">gtf_url</span>
    <span class="n">final_chrom_sizes_url</span> <span class="o">=</span> <span class="n">chrom_sizes_url</span>
    <span class="n">final_fasta_url</span> <span class="o">=</span> <span class="n">fasta_url</span>

    <span class="c1"># Known defaults for mouse mm10</span>
    <span class="k">if</span> <span class="n">species</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;mouse&quot;</span> <span class="ow">and</span> <span class="n">assembly</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;mm10&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">final_gtf_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">final_gtf_url</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/gencode/&quot;</span>
                <span class="s2">&quot;Gencode_mouse/release_M18/gencode.vM18.basic.annotation.gtf.gz&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">final_chrom_sizes_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">final_chrom_sizes_url</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;https://hgdownload.cse.ucsc.edu/goldenpath/mm10/bigZips/mm10.chrom.sizes&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">final_fasta_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">final_fasta_url</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;https://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.fa.gz&quot;</span>
            <span class="p">)</span>

    <span class="c1"># Known defaults for human hg38</span>
    <span class="k">elif</span> <span class="n">species</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;human&quot;</span> <span class="ow">and</span> <span class="n">assembly</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;hg38&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">final_gtf_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">final_gtf_url</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/gencode/&quot;</span>
                <span class="s2">&quot;Gencode_human/release_47/gencode.v47.primary_assembly.basic.annotation.gtf.gz&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">final_chrom_sizes_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">final_chrom_sizes_url</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.chrom.sizes&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">final_fasta_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">final_fasta_url</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.fa.gz&quot;</span>
            <span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Unknown assembly =&gt; user must provide or raise an error if any is None</span>
        <span class="k">if</span> <span class="n">final_gtf_url</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">final_chrom_sizes_url</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">final_fasta_url</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Unknown assembly &#39;</span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">&#39; for species=&#39;</span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">&#39;. &quot;</span>
                <span class="s2">&quot;Please provide gtf_url, chrom_sizes_url, and fasta_url in config.&quot;</span>
            <span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">final_gtf_url</span><span class="p">,</span> <span class="n">final_chrom_sizes_url</span><span class="p">,</span> <span class="n">final_fasta_url</span><span class="p">)</span></div>


<div class="viewcode-block" id="download_genome_references">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.download.download_genome_references">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">download_genome_references</span><span class="p">(</span>
    <span class="n">genome_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">assembly</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gtf_url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">chrom_sizes_url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">fasta_url</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download reference genome files (GTF, chrom.sizes, FASTA) for a given species and assembly.</span>

<span class="sd">    1. Resolve the URLs from user input or known defaults for recognized combos (mouse/mm10, human/hg38).</span>
<span class="sd">    2. Download each file if not already present.</span>
<span class="sd">    3. Decompress .gz files for GTF and FASTA.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    genome_dir : pathlib.Path</span>
<span class="sd">        Directory where the reference files will be downloaded.</span>
<span class="sd">    species : str</span>
<span class="sd">        Species name (e.g. &quot;mouse&quot;, &quot;human&quot;).</span>
<span class="sd">    assembly : str</span>
<span class="sd">        Genome assembly version (e.g. &quot;mm10&quot;, &quot;hg38&quot;).</span>
<span class="sd">    gtf_url : str, optional</span>
<span class="sd">        GTF file URL to override defaults. If None, use known default (if available).</span>
<span class="sd">    chrom_sizes_url : str, optional</span>
<span class="sd">        Chromosome sizes file URL to override defaults. If None, use known default (if available).</span>
<span class="sd">    fasta_url : str, optional</span>
<span class="sd">        FASTA file URL to override defaults. If None, use known default (if available).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        Files are downloaded into `genome_dir`. If a file already exists, no new download occurs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genome_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># 1) Resolve the final URLs based on species/assembly + user overrides</span>
    <span class="n">final_gtf_url</span><span class="p">,</span> <span class="n">final_chrom_sizes_url</span><span class="p">,</span> <span class="n">final_fasta_url</span> <span class="o">=</span> <span class="n">resolve_genome_urls</span><span class="p">(</span>
        <span class="n">species</span><span class="p">,</span> <span class="n">assembly</span><span class="p">,</span> <span class="n">gtf_url</span><span class="p">,</span> <span class="n">chrom_sizes_url</span><span class="p">,</span> <span class="n">fasta_url</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Using genome references for species=&#39;</span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">&#39;, assembly=&#39;</span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">&#39;.</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot;GTF: </span><span class="si">{</span><span class="n">final_gtf_url</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot;Chrom.sizes: </span><span class="si">{</span><span class="n">final_chrom_sizes_url</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="sa">f</span><span class="s2">&quot;FASTA: </span><span class="si">{</span><span class="n">final_fasta_url</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Decide on local filenames</span>
    <span class="n">gtf_gz</span> <span class="o">=</span> <span class="n">genome_dir</span> <span class="o">/</span> <span class="s2">&quot;annotation.gtf.gz&quot;</span>
    <span class="n">gtf_final</span> <span class="o">=</span> <span class="n">genome_dir</span> <span class="o">/</span> <span class="s2">&quot;annotation.gtf&quot;</span>
    <span class="n">chrom_sizes_path</span> <span class="o">=</span> <span class="n">genome_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">.chrom.sizes&quot;</span>
    <span class="n">fasta_gz</span> <span class="o">=</span> <span class="n">genome_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">.fa.gz&quot;</span>
    <span class="n">fasta_final</span> <span class="o">=</span> <span class="n">genome_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">assembly</span><span class="si">}</span><span class="s2">.fa&quot;</span>

    <span class="c1"># 2) GTF</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">gtf_final</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">download_file</span><span class="p">(</span><span class="n">final_gtf_url</span><span class="p">,</span> <span class="n">gtf_gz</span><span class="p">)</span>
        <span class="n">unzip_gz</span><span class="p">(</span><span class="n">gtf_gz</span><span class="p">,</span> <span class="n">remove_input</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># 3) chrom sizes</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">chrom_sizes_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">download_file</span><span class="p">(</span><span class="n">final_chrom_sizes_url</span><span class="p">,</span> <span class="n">chrom_sizes_path</span><span class="p">)</span>

    <span class="c1"># 4) FASTA</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">fasta_final</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">download_file</span><span class="p">(</span><span class="n">final_fasta_url</span><span class="p">,</span> <span class="n">fasta_gz</span><span class="p">)</span>
        <span class="n">unzip_gz</span><span class="p">(</span><span class="n">fasta_gz</span><span class="p">,</span> <span class="n">remove_input</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reference files are ready in </span><span class="si">{</span><span class="n">genome_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>

</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">scDoRI</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../preprocessing_pipeline.html">Preprocessing Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../scdori.html">scDORI training</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks.html">Notebook Examples</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Manu Saraswat.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
    </div>

    

    
  </body>
</html>