
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Loading and preparing data for training and model initialisation &#8212; scDoRI  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/training';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">scDoRI  documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../api_reference.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saraswatmanu/scDoRI" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../api_reference.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/saraswatmanu/scDoRI" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"></div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Loading and preparing data for training and model initialisation</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <p>Notebook for scDoRI model training and downstream analysis</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">torch.utils.data</span><span class="w"> </span><span class="kn">import</span> <span class="n">DataLoader</span><span class="p">,</span> <span class="n">TensorDataset</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.model_selection</span><span class="w"> </span><span class="kn">import</span> <span class="n">train_test_split</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">scdori</span><span class="w"> </span><span class="kn">import</span> <span class="n">config</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.data_io</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_scdori_inputs</span><span class="p">,</span> <span class="n">save_model_weights</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">set_seed</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">scDoRI</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.train_scdori</span><span class="w"> </span><span class="kn">import</span> <span class="n">train_scdori_phases</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.train_grn</span><span class="w"> </span><span class="kn">import</span> <span class="n">train_model_grn</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">initialize_scdori_parameters</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">importlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">reload</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>

<span class="k">for</span> <span class="n">module_name</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">modules</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="k">if</span> <span class="n">module_name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;scdori&quot;</span><span class="p">):</span>
        <span class="n">reload</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">modules</span><span class="p">[</span><span class="n">module_name</span><span class="p">])</span>
</pre></div>
</div>
</div>
<section id="Loading-and-preparing-data-for-training-and-model-initialisation">
<h1>Loading and preparing data for training and model initialisation<a class="headerlink" href="#Loading-and-preparing-data-for-training-and-model-initialisation" title="Link to this heading">#</a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">logging_level</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting scDoRI pipeline with integrated GRN.&quot;</span><span class="p">)</span>
<span class="n">set_seed</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">random_seed</span><span class="p">)</span>

<span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda:0&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Using device: </span><span class="si">{</span><span class="n">device</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:__main__:Starting scDoRI pipeline with integrated GRN.
INFO:scdori.utils:Random seed set to 200.
INFO:__main__:Using device: cuda:0
</pre></div></div>
</div>
<section id="1.-load-data">
<h2>1. load data<a class="headerlink" href="#1.-load-data" title="Link to this heading">#</a></h2>
<p>uses the path specified in config file to load processed RNA and ATAC anndata as well as precomputed insilico-chipseq matrix and peak-gene distances</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">atac_metacell</span><span class="p">,</span> <span class="n">gene_peak_dist</span><span class="p">,</span> <span class="n">insilico_act</span><span class="p">,</span> <span class="n">insilico_rep</span> <span class="o">=</span> <span class="n">load_scdori_inputs</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
<span class="n">gene_peak_fixed</span> <span class="o">=</span> <span class="n">gene_peak_dist</span><span class="o">.</span><span class="n">clone</span><span class="p">()</span>
<span class="n">gene_peak_fixed</span><span class="p">[</span><span class="n">gene_peak_fixed</span> <span class="o">&gt;</span><span class="mi">0</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>  <span class="c1"># mask for peak-gene links based on distance</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.data_io:Loading RNA from /data/saraswat/new_metacells/data_gastrulation_single_cell/generated/rna_processed.h5ad
INFO:scdori.data_io:Loading ATAC from /data/saraswat/new_metacells/data_gastrulation_single_cell/generated/atac_processed.h5ad
INFO:scdori.data_io:Loading gene-peak dist from /data/saraswat/new_metacells/data_gastrulation_single_cell/generated/gene_peak_distance_exp.npy
INFO:scdori.data_io:Loading insilico embeddings from /data/saraswat/new_metacells/data_gastrulation_single_cell/generated/insilico_chipseq_act.npy &amp; /data/saraswat/new_metacells/data_gastrulation_single_cell/generated/insilico_chipseq_rep.npy
</pre></div></div>
</div>
</section>
<section id="2.-computing-indices-of-genes-which-are-TFs-and-setting-number-of-cells-per-metacell-(-set-to-1-for-single-cell-data)">
<h2>2. computing indices of genes which are TFs and setting number of cells per metacell ( set to 1 for single cell data)<a class="headerlink" href="#2.-computing-indices-of-genes-which-are-TFs-and-setting-number-of-cells-per-metacell-(-set-to-1-for-single-cell-data)" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># computing indices of genes which are TFs and setting number of cells per metacell ( set to 1 for single cell data)</span>
<span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;num_cells&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
<span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;index_int&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">tf_indices</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_type</span><span class="o">==</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index_int</span><span class="o">.</span><span class="n">values</span>
<span class="n">num_cells</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">num_cells</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<br/></pre></div>
</div>
</div>
</section>
<section id="3.-onehot-encoding-the-batch-column-for-entire-dataset">
<h2>3. onehot encoding the batch column for entire dataset<a class="headerlink" href="#3.-onehot-encoding-the-batch-column-for-entire-dataset" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">batch_col</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">batch_col</span>
<span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">atac_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">batch_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="c1"># obtaining onehot encoding for technical batch,</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">OneHotEncoder</span>
<span class="n">enc</span> <span class="o">=</span> <span class="n">OneHotEncoder</span><span class="p">(</span><span class="n">handle_unknown</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="n">enc</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="p">)</span>

<span class="n">onehot_batch</span> <span class="o">=</span> <span class="n">enc</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
<span class="n">enc</span><span class="o">.</span><span class="n">categories_</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[array([&#39;E7.5_rep1&#39;, &#39;E7.5_rep2&#39;, &#39;E7.75_rep1&#39;, &#39;E8.0_rep1&#39;, &#39;E8.0_rep2&#39;,
        &#39;E8.5_CRISPR_T_KO&#39;, &#39;E8.5_CRISPR_T_WT&#39;, &#39;E8.5_rep1&#39;, &#39;E8.5_rep2&#39;,
        &#39;E8.75_rep1&#39;, &#39;E8.75_rep2&#39;], dtype=object)]
</pre></div></div>
</div>
</section>
<section id="4.-making-train-and-evaluation-datasets">
<h2>4. making train and evaluation datasets<a class="headerlink" href="#4.-making-train-and-evaluation-datasets" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 2) Make small train/test sets</span>
<span class="n">n_cells</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">n_obs</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n_cells</span><span class="p">)</span>
<span class="n">train_idx</span><span class="p">,</span> <span class="n">eval_idx</span> <span class="o">=</span> <span class="n">train_test_split</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">test_size</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">train_dataset</span> <span class="o">=</span> <span class="n">TensorDataset</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">train_idx</span><span class="p">))</span>
<span class="n">train_loader</span>  <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span><span class="n">train_dataset</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">batch_size_cell</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">eval_dataset</span>  <span class="o">=</span> <span class="n">TensorDataset</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">eval_idx</span><span class="p">))</span>
<span class="n">eval_loader</span>   <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span><span class="n">eval_dataset</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">batch_size_cell</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="5.-Build-scDoRI-model-using-parameters-from-config-file">
<h2>5. Build scDoRI model using parameters from config file<a class="headerlink" href="#5.-Build-scDoRI-model-using-parameters-from-config-file" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">num_genes</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">n_vars</span>
<span class="n">num_peaks</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">n_vars</span>

<span class="n">num_tfs</span>   <span class="o">=</span> <span class="n">insilico_act</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

<span class="n">num_batches</span> <span class="o">=</span> <span class="n">onehot_batch</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.models</span><span class="w"> </span><span class="kn">import</span> <span class="n">scDoRI</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">scDoRI</span><span class="p">(</span>
    <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">,</span>
    <span class="n">num_genes</span><span class="o">=</span><span class="n">num_genes</span><span class="p">,</span>
    <span class="n">num_peaks</span><span class="o">=</span><span class="n">num_peaks</span><span class="p">,</span>
    <span class="n">num_tfs</span><span class="o">=</span><span class="n">num_tfs</span><span class="p">,</span>
    <span class="n">num_topics</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">num_topics</span><span class="p">,</span>
    <span class="n">num_batches</span><span class="o">=</span><span class="n">num_batches</span><span class="p">,</span>
    <span class="n">dim_encoder1</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">dim_encoder1</span><span class="p">,</span>
    <span class="n">dim_encoder2</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">dim_encoder2</span>
<span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="6.-initialising-scDoRI-model-with-precomputed-matrices-and-setting-gradients">
<h2>6. initialising scDoRI model with precomputed matrices and setting gradients<a class="headerlink" href="#6.-initialising-scDoRI-model-with-precomputed-matrices-and-setting-gradients" title="Link to this heading">#</a></h2>
<p>initialising with precomputed insilico-chipseq matrices and distance dependent peak-gene links</p>
<p>also setting corresponding gradients for TF-gene links to False as they are not updated in Phase 1 of training</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">initialize_scdori_parameters</span><span class="p">(</span>
    <span class="n">model</span><span class="p">,</span>
    <span class="n">gene_peak_dist</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">),</span> <span class="n">gene_peak_fixed</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">),</span>
    <span class="n">insilico_act</span><span class="o">=</span><span class="n">insilico_act</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">),</span>
    <span class="n">insilico_rep</span><span class="o">=</span><span class="n">insilico_rep</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">),</span><span class="n">phase</span><span class="o">=</span><span class="s2">&quot;warmup&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scDoRI parameters (peak-gene distance &amp; TF binding) initialized and relevant parameters frozen.
</pre></div></div>
</div>
</section>
</section>
<section id="Train-Phase-1-of-scDoRI-model">
<h1>Train Phase 1 of scDoRI model<a class="headerlink" href="#Train-Phase-1-of-scDoRI-model" title="Link to this heading">#</a></h1>
<p>here topics are inferred using reconstruction of ATAC peaks (module 1), reconstruction of RNA from predicted ATAC (module 2) and reconstruction of TF expression (module 3)</p>
<p>Warmup start is used where only module 1 and module 3 are trained for some initial epochs before adding module 2</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">train_scdori_phases</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">train_loader</span><span class="p">,</span> <span class="n">eval_loader</span><span class="p">,</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">atac_metacell</span><span class="p">,</span><span class="n">num_cells</span><span class="p">,</span> <span class="n">tf_indices</span><span class="p">,</span> <span class="n">onehot_batch</span><span class="p">,</span><span class="n">config</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># saving the model weight correspoinding to final epoch where model stopped training</span>
<span class="n">save_model_weights</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">weights_folder_scdori</span><span class="p">),</span> <span class="s2">&quot;scdori_final&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Train-Phase-2-of-scDoRI-model">
<h1>Train Phase 2 of scDoRI model<a class="headerlink" href="#Train-Phase-2-of-scDoRI-model" title="Link to this heading">#</a></h1>
<p>here activator and repressor TF-gene links per topic are inferred (module 4)</p>
<p>optionally the encoder and other model parameters from module 1,2,3 are frozen for stability</p>
<section id="7.-Load-best-checkpoint-from-Phase-1">
<h2>7. Load best checkpoint from Phase 1<a class="headerlink" href="#7.-Load-best-checkpoint-from-Phase-1" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Phase 2</span>


<span class="c1"># loading the best checkpoint from Phase 1</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.downstream</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_best_model</span>

<span class="n">model</span> <span class="o">=</span> <span class="n">load_best_model</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">weights_folder_scdori</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;best_scdori_best_eval.pth&quot;</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="8.-Set-gradients-for-Phase-2-training">
<h2>8. Set gradients for Phase 2 training<a class="headerlink" href="#8.-Set-gradients-for-Phase-2-training" title="Link to this heading">#</a></h2>
<p>TF-gene links are learnt in this step</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><br/><span></span><span class="n">initialize_scdori_parameters</span><span class="p">(</span>
    <span class="n">model</span><span class="p">,</span>
    <span class="n">gene_peak_dist</span><span class="p">,</span> <span class="n">gene_peak_fixed</span><span class="p">,</span>
    <span class="n">insilico_act</span><span class="o">=</span><span class="n">insilico_act</span><span class="p">,</span>
    <span class="n">insilico_rep</span><span class="o">=</span><span class="n">insilico_rep</span><span class="p">,</span><span class="n">phase</span><span class="o">=</span><span class="s2">&quot;grn&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="9.-Phase-2-training-and-saving-model-weights">
<h2>9. Phase 2 training and saving model weights<a class="headerlink" href="#9.-Phase-2-training-and-saving-model-weights" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># train Phase 2 of scDoRI model, TF-gene links are learnt in this phase and used to reconstruct gene-expression profiles</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">train_model_grn</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">train_loader</span><span class="p">,</span> <span class="n">eval_loader</span><span class="p">,</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">atac_metacell</span><span class="p">,</span><span class="n">num_cells</span><span class="p">,</span> <span class="n">tf_indices</span><span class="p">,</span> <span class="n">onehot_batch</span><span class="p">,</span><span class="n">config</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># saving the model weight correspoinding to final epoch where model stopped training</span>
<span class="n">save_model_weights</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">weights_folder_grn</span><span class="p">),</span> <span class="s2">&quot;scdori_final&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Downstream-analysis">
<h1>Downstream analysis<a class="headerlink" href="#Downstream-analysis" title="Link to this heading">#</a></h1>
<p>scDoRI supports a comprehensive suite of downstream analyses for single-cell multiome RNA-ATAC data. These include dimensionality reduction using latent topics, identification of gene and peak programs associated with each topic, inference of enhancer–gene interactions, and construction of topic-specific transcription factor–gene regulatory networks (GRNs).</p>
<p>We demonstrate these capabilities using the mouse gastrulation dataset from <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2022.06.15.496239v1">https://www.biorxiv.org/content/10.1101/2022.06.15.496239v1</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[226]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">scdori.downstream</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
        <span class="n">load_best_model</span><span class="p">,</span>

        <span class="n">compute_neighbors_umap</span><span class="p">,</span>
        <span class="n">compute_topic_peak_umap</span><span class="p">,</span><span class="n">compute_topic_gene_matrix</span><span class="p">,</span>
        <span class="n">compute_atac_grn_activator_with_significance</span><span class="p">,</span>
        <span class="n">compute_atac_grn_repressor_with_significance</span><span class="p">,</span>
        <span class="n">compute_significant_grn</span><span class="p">,</span>
        <span class="n">visualize_downstream_targets</span><span class="p">,</span>
        <span class="n">plot_topic_activation_heatmap</span><span class="p">,</span>
        <span class="n">get_top_activators_per_topic</span><span class="p">,</span>
        <span class="n">get_top_repressor_per_topic</span><span class="p">,</span>
        <span class="n">compute_activator_tf_activity_per_cell</span><span class="p">,</span>
        <span class="n">compute_repressor_tf_activity_per_cell</span><span class="p">,</span><span class="n">save_regulons</span>
    <span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.evaluation</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_latent_topics</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">MinMaxScaler</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scdori.train_grn</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_tf_expression</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[227]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">importlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">reload</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>

<span class="k">for</span> <span class="n">module_name</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">modules</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="k">if</span> <span class="n">module_name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;scdori&quot;</span><span class="p">):</span>
        <span class="n">reload</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">modules</span><span class="p">[</span><span class="n">module_name</span><span class="p">])</span>
</pre></div>
</div>
</div>
<section id="10.-Load-best-checkpoint-model">
<h2>10. Load best checkpoint model<a class="headerlink" href="#10.-Load-best-checkpoint-model" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">load_best_model</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">weights_folder_grn</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;best_scdori_best_eval.pth&quot;</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:Loaded best model weights from weights_directory_grn/best_scdori_best_eval.pth
</pre></div></div>
</div>
</section>
<section id="11.-Computing-and-visualising-latent-topic-activity-per-cell">
<h2>11. Computing and visualising latent topic activity per cell<a class="headerlink" href="#11.-Computing-and-visualising-latent-topic-activity-per-cell" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># creating dataloader for all cells</span>
<span class="n">n_cells</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">n_obs</span>
<span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n_cells</span><span class="p">)</span>

<span class="n">all_dataset</span> <span class="o">=</span> <span class="n">TensorDataset</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">indices</span><span class="p">))</span>
<span class="n">all_dataset_loader</span>  <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span><span class="n">all_dataset</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="n">config</span><span class="o">.</span><span class="n">batch_size_cell_prediction</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># get scDoRI latent embedding (topics)</span>
<span class="n">scdori_latent</span> <span class="o">=</span> <span class="n">get_latent_topics</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">all_dataset_loader</span><span class="p">,</span><span class="n">rna_metacell</span><span class="p">,</span><span class="n">atac_metacell</span><span class="p">,</span><span class="n">num_cells</span><span class="p">,</span><span class="n">tf_indices</span><span class="p">,</span><span class="n">onehot_batch</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Extracting latent topics: 100%|██████████| 112/112 [01:26&lt;00:00,  1.29it/s]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># adding scDoRI embedding to the anndata object</span>
<span class="n">rna_metacell</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_scdori&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scdori_latent</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## adding color palette</span>
<span class="n">colPalette_celltypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#532C8A&#39;</span><span class="p">,</span> <span class="s1">&#39;#c19f70&#39;</span><span class="p">,</span> <span class="s1">&#39;#f9decf&#39;</span><span class="p">,</span><span class="s1">&#39;#c9a997&#39;</span><span class="p">,</span> <span class="s1">&#39;#B51D8D&#39;</span><span class="p">,</span> <span class="s1">&#39;#3F84AA&#39;</span><span class="p">,</span> <span class="s1">&#39;#9e6762&#39;</span><span class="p">,</span> <span class="s1">&#39;#354E23&#39;</span><span class="p">,</span> <span class="s1">&#39;#F397C0&#39;</span><span class="p">,</span>
                        <span class="s1">&#39;#ff891c&#39;</span><span class="p">,</span> <span class="s1">&#39;#635547&#39;</span><span class="p">,</span> <span class="s1">&#39;#C72228&#39;</span><span class="p">,</span> <span class="s1">&#39;#f79083&#39;</span><span class="p">,</span> <span class="s1">&#39;#EF4E22&#39;</span><span class="p">,</span> <span class="s1">&#39;#989898&#39;</span><span class="p">,</span> <span class="s1">&#39;#7F6874&#39;</span><span class="p">,</span> <span class="s1">&#39;#8870ad&#39;</span><span class="p">,</span> <span class="s1">&#39;#647a4f&#39;</span><span class="p">,</span>
 <span class="s1">&#39;#EF5A9D&#39;</span><span class="p">,</span> <span class="s1">&#39;#FBBE92&#39;</span><span class="p">,</span> <span class="s1">&#39;#139992&#39;</span><span class="p">,</span> <span class="s1">&#39;#cc7818&#39;</span><span class="p">,</span> <span class="s1">&#39;#DFCDE4&#39;</span><span class="p">,</span> <span class="s1">&#39;#8EC792&#39;</span><span class="p">,</span> <span class="s1">&#39;#C594BF&#39;</span><span class="p">,</span> <span class="s1">&#39;#C3C388&#39;</span><span class="p">,</span> <span class="s1">&#39;#0F4A9C&#39;</span><span class="p">,</span> <span class="s1">&#39;#FACB12&#39;</span><span class="p">,</span> <span class="s1">&#39;#8DB5CE&#39;</span><span class="p">,</span> <span class="s1">&#39;#1A1A1A&#39;</span><span class="p">,</span>
 <span class="s1">&#39;#C9EBFB&#39;</span><span class="p">,</span> <span class="s1">&#39;#DABE99&#39;</span><span class="p">,</span> <span class="s1">&#39;#65A83E&#39;</span><span class="p">,</span> <span class="s1">&#39;#005579&#39;</span><span class="p">,</span> <span class="s1">&#39;#CDE088&#39;</span><span class="p">,</span> <span class="s1">&#39;#f7f79e&#39;</span><span class="p">,</span> <span class="s1">&#39;#F6BFCB&#39;</span><span class="p">]</span>

<span class="n">cell_type_sorted</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)))</span>

<span class="n">color_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">cell_type_sorted</span><span class="p">,</span><span class="n">colPalette_celltypes</span><span class="p">))</span>

<span class="n">col_sorted</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">cell_type_sorted</span><span class="p">):</span>
    <span class="n">col_sorted</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">color_dict</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

<span class="n">rna_metacell</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;celltype_colors&#39;</span><span class="p">]</span><span class="o">=</span> <span class="n">col_sorted</span>
<span class="c1">#rna_metacell.uns[&#39;celltype_plot_colors&#39;]= col_sorted</span>
<span class="c1">#rna_ad_tl.uns[&#39;celltype_colors&#39;]= col_sorted</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># computing neighbourhood graph and UMAP based on scDoRI embedding, UMAP parameters can be set in config file</span>
<span class="n">compute_neighbors_umap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">rep_key</span><span class="o">=</span><span class="s2">&quot;X_scdori&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Computing neighbors + UMAP on scDoRI latent ===
INFO:scdori.downstream:Done. UMAP stored in rna_anndata.obsm[&#39;X_umap&#39;].
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="c1"># visualing cell-types on scDoRI computed UMAP</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">0.3</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">200</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">],</span><span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">outline_color</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="s1">&#39;black&#39;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_39_0.png" src="../_images/notebooks_training_39_0.png" />
</div>
</div>
</section>
<section id="12.-Computing-average-topic-activation-in-different-celltypes/groups">
<h2>12. Computing average topic activation in different celltypes/groups<a class="headerlink" href="#12.-Computing-average-topic-activation-in-different-celltypes/groups" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="n">df_topic_celltype</span> <span class="o">=</span> <span class="n">plot_topic_activation_heatmap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">groupby_key</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">],</span> <span class="n">aggregation</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_41_0.png" src="../_images/notebooks_training_41_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[333]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">df_topic_celltype</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[333]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: ylabel=&#39;Count&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_42_1.png" src="../_images/notebooks_training_42_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[337]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># removing topics not active highly in any of the celltypes</span>
<span class="n">select_topics</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Topic_&quot;</span> <span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">df_topic_celltype</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">&gt;</span><span class="mf">0.06</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[338]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">select_topics</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[338]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
25
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[339]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># masking low activations and removing topics not active</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">df_topic_celltype</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">select_topics</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Greens&quot;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[339]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.matrix.ClusterGrid at 0x7f6f65d2ab90&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_45_1.png" src="../_images/notebooks_training_45_1.png" />
</div>
</div>
</section>
<section id="13.-single-cell-level-visualisation-of-scDoRI-Topics">
<h2>13. single cell level visualisation of scDoRI Topics<a class="headerlink" href="#13.-single-cell-level-visualisation-of-scDoRI-Topics" title="Link to this heading">#</a></h2>
<p>In Erythroid Trajectory from Blood Progenitors to Mature Erythroids: We can see that Topic 26 is active progenitors whereas topic 4 and 9 capture a continuum of erythroid and progenitor programs along the trajectory</p>
<p>In transition of bipotent progenitors NMP (Meuromesodermal progenitors) to spinal cord or mesodermal lineage: We can see that Topic 34 and 7 capture transition to mesoderm and spinal cord respectively</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">num_topics</span><span class="p">):</span>
    <span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Topic_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="p">)]</span> <span class="o">=</span> <span class="n">scdori_latent</span><span class="p">[:,</span><span class="n">k</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>



<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">150</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span><span class="s2">&quot;Topic_26&quot;</span><span class="p">,</span><span class="s2">&quot;Topic_4&quot;</span><span class="p">,</span> <span class="s2">&quot;Topic_9&quot;</span><span class="p">],</span><span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">outline_color</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="s1">&#39;black&#39;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_48_0.png" src="../_images/notebooks_training_48_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># visualising NMP transitions to Mesoderm [Topic 34] and Spinal Cord [Topic 7] Trajectory</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>

<span class="n">celltype_plot_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Mixed_mesoderm&#39;</span><span class="p">,</span>
 <span class="s1">&#39;Somitic_mesoderm&#39;</span><span class="p">,</span> <span class="s1">&#39;Nascent_mesoderm&#39;</span><span class="p">,</span><span class="s1">&#39;Intermediate_mesoderm&#39;</span><span class="p">,</span><span class="s1">&#39;Forebrain_Midbrain_Hindbrain&#39;</span><span class="p">,</span>
 <span class="s1">&#39;Spinal_cord&#39;</span><span class="p">,</span> <span class="s1">&#39;Caudal_Mesoderm&#39;</span><span class="p">,</span><span class="s1">&#39;NMP&#39;</span><span class="p">,</span> <span class="s1">&#39;ExE_mesoderm&#39;</span><span class="p">,</span> <span class="s1">&#39;Allantois&#39;</span><span class="p">]</span>
<span class="n">adata_plot</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="p">[</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">celltype</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">celltype_plot_list</span><span class="p">),:]</span>

<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">100</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_plot</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span><span class="s2">&quot;Topic_34&quot;</span><span class="p">,</span> <span class="s2">&quot;Topic_7&quot;</span><span class="p">],</span><span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">outline_color</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="s1">&#39;black&#39;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_49_0.png" src="../_images/notebooks_training_49_0.png" />
</div>
</div>
</section>
<section id="14.-Computing-Top-genes-per-topic">
<h2>14. Computing Top genes per topic<a class="headerlink" href="#14.-Computing-Top-genes-per-topic" title="Link to this heading">#</a></h2>
<p>this can be used for further analysis such as gene-set enrichment</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">topic_gene_embedding</span> <span class="o">=</span> <span class="n">compute_topic_gene_matrix</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:Done. computing topic gene matrix shape =&gt; torch.Size([40, 4000])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[237]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">topic_gene_embedding</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/data/saraswat/miniforge3/envs/scdori_test_env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
  warnings.warn(msg)
/data/saraswat/miniforge3/envs/scdori_test_env/lib/python3.10/site-packages/seaborn/matrix.py:560: UserWarning: Clustering large matrix with scipy. Installing `fastcluster` may give better performance.
  warnings.warn(msg)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[237]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.matrix.ClusterGrid at 0x7f6e88338460&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_52_2.png" src="../_images/notebooks_training_52_2.png" />
</div>
</div>
</section>
<section id="15.-Performing-gene-set-enrichment-analysis-on-each-topic">
<h2>15. Performing gene-set enrichment analysis on each topic<a class="headerlink" href="#15.-Performing-gene-set-enrichment-analysis-on-each-topic" title="Link to this heading">#</a></h2>
<p>adapted from <a class="reference external" href="https://decoupler-py.readthedocs.io/en/latest/notebooks/msigdb.html#MSigDB-gene-sets">https://decoupler-py.readthedocs.io/en/latest/notebooks/msigdb.html#MSigDB-gene-sets</a></p>
<p>users can play around with other gene sets of their choice</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!pip install decoupler</span>
<span class="c1">### caution! this can create dependency issues in the conda environment</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[262]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="ch">#!pip install liana</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[258]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">decoupler</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">dc</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[265]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># anndata with topic gene values</span>
<span class="n">adata_gene</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">topic_gene_embedding</span><span class="p">)</span>
<span class="n">adata_gene</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Topic_&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">num_topics</span><span class="p">)]</span>
<span class="n">adata_gene</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">]</span> <span class="c1"># dirty and incorrect hack to convert mouse gene names to human</span>
<span class="n">adata_gene</span><span class="o">.</span><span class="n">raw</span><span class="o">=</span><span class="n">adata_gene</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[291]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># using msigdb</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_resource</span><span class="p">(</span><span class="s1">&#39;MSigDB&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[292]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">msigdb</span><span class="p">[</span><span class="s1">&#39;collection&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[292]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;StringArray&gt;
[                       &#39;immunesigdb&#39;,                  &#39;tf_targets_legacy&#39;,
                         &#39;positional&#39;,               &#39;cell_type_signatures&#39;,
              &#39;go_cellular_component&#39;, &#39;chemical_and_genetic_perturbations&#39;,
                &#39;mirna_targets_mirdb&#39;,                   &#39;vaccine_response&#39;,
                     &#39;cancer_modules&#39;,                  &#39;reactome_pathways&#39;,
                    &#39;tf_targets_gtrf&#39;,              &#39;go_biological_process&#39;,
              &#39;go_molecular_function&#39;,               &#39;oncogenic_signatures&#39;,
                           &#39;hallmark&#39;,                      &#39;kegg_pathways&#39;,
                       &#39;pid_pathways&#39;,           &#39;human_phenotype_ontology&#39;,
                       &#39;wikipathways&#39;,          &#39;cancer_gene_neighborhoods&#39;,
               &#39;mirna_targets_legacy&#39;,                  &#39;biocarta_pathways&#39;]
Length: 22, dtype: string
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[293]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># using msigdb celltype signature set</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_resource</span><span class="p">(</span><span class="s1">&#39;MSigDB&#39;</span><span class="p">)</span>
<span class="c1"># Filter by hallmark</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="n">msigdb</span><span class="p">[</span><span class="s1">&#39;collection&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;cell_type_signatures&#39;</span><span class="p">]</span>

<span class="c1"># Remove duplicated entries</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="o">~</span><span class="n">msigdb</span><span class="o">.</span><span class="n">duplicated</span><span class="p">([</span><span class="s1">&#39;geneset&#39;</span><span class="p">,</span> <span class="s1">&#39;genesymbol&#39;</span><span class="p">])]</span>
<span class="n">msigdb</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[293]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>genesymbol</th>
      <th>collection</th>
      <th>geneset</th>
    </tr>
    <tr>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>A1BG</td>
      <td>cell_type_signatures</td>
      <td>GAO_LARGE_INTESTINE_ADULT_CI_MESENCHYMAL_CELLS</td>
    </tr>
    <tr>
      <th>7</th>
      <td>A1BG</td>
      <td>cell_type_signatures</td>
      <td>HE_LIM_SUN_FETAL_LUNG_C1_PROXIMAL_BASAL_CELL</td>
    </tr>
    <tr>
      <th>9</th>
      <td>A1BG</td>
      <td>cell_type_signatures</td>
      <td>HE_LIM_SUN_FETAL_LUNG_C2_PRE_PDC_DC5_CELL</td>
    </tr>
    <tr>
      <th>12</th>
      <td>A1BG</td>
      <td>cell_type_signatures</td>
      <td>HE_LIM_SUN_FETAL_LUNG_C2_APOE_POS_M2_MACROPHAG...</td>
    </tr>
    <tr>
      <th>14</th>
      <td>A1BG</td>
      <td>cell_type_signatures</td>
      <td>GAUTAM_EYE_CORNEA_MELANOCYTES</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5521812</th>
      <td>ZYX</td>
      <td>cell_type_signatures</td>
      <td>BUSSLINGER_DUODENAL_IMMUNE_CELLS</td>
    </tr>
    <tr>
      <th>5521814</th>
      <td>ZYX</td>
      <td>cell_type_signatures</td>
      <td>MANNO_MIDBRAIN_NEUROTYPES_HPERIC</td>
    </tr>
    <tr>
      <th>5521818</th>
      <td>ZYX</td>
      <td>cell_type_signatures</td>
      <td>TRAVAGLINI_LUNG_BRONCHIAL_VESSEL_2_CELL</td>
    </tr>
    <tr>
      <th>5521916</th>
      <td>ZZEF1</td>
      <td>cell_type_signatures</td>
      <td>GAO_ESOPHAGUS_25W_C4_FGFR1HIGH_EPITHELIAL_CELLS</td>
    </tr>
    <tr>
      <th>5522071</th>
      <td>ZZZ3</td>
      <td>cell_type_signatures</td>
      <td>MURARO_PANCREAS_BETA_CELL</td>
    </tr>
  </tbody>
</table>
<p>192576 rows × 3 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[295]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dc</span><span class="o">.</span><span class="n">run_ora</span><span class="p">(</span>
    <span class="n">mat</span><span class="o">=</span><span class="n">adata_gene</span><span class="p">,</span>
    <span class="n">net</span><span class="o">=</span><span class="n">msigdb</span><span class="p">,</span>
    <span class="n">source</span><span class="o">=</span><span class="s1">&#39;geneset&#39;</span><span class="p">,</span>
    <span class="n">target</span><span class="o">=</span><span class="s1">&#39;genesymbol&#39;</span><span class="p">,</span>
    <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running ora on mat with 40 samples and 4000 targets for 775 sources.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "67e5034102cf4da7bd152590c843e220", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[296]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">acts</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_acts</span><span class="p">(</span><span class="n">adata_gene</span><span class="p">,</span> <span class="n">obsm_key</span><span class="o">=</span><span class="s1">&#39;ora_estimate&#39;</span><span class="p">)</span>

<span class="c1"># We need to remove inf and set them to the maximum value observed</span>
<span class="n">acts_v</span> <span class="o">=</span> <span class="n">acts</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="n">max_e</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">acts_v</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">acts_v</span><span class="p">)])</span>
<span class="n">acts</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">acts</span><span class="o">.</span><span class="n">X</span><span class="p">)]</span> <span class="o">=</span> <span class="n">max_e</span>

<span class="n">acts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[296]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 40 × 775
    obsm: &#39;ora_estimate&#39;, &#39;ora_pvals&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[298]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot top celltype signatures per topic</span>
<span class="n">df_acts</span> <span class="o">=</span> <span class="n">acts</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
<span class="n">top_programs_per_topic</span> <span class="o">=</span> <span class="n">df_acts</span><span class="o">.</span><span class="n">idxmax</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">unique_top_programs</span> <span class="o">=</span> <span class="n">top_programs_per_topic</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">df_topic_program</span> <span class="o">=</span> <span class="n">df_acts</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">unique_top_programs</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[352]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">df_topic_program</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">select_topics</span><span class="p">],</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span><span class="mi">30</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[352]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.matrix.ClusterGrid at 0x7f68f63ef610&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_64_1.png" src="../_images/notebooks_training_64_1.png" />
</div>
</div>
<p>we can see enrichments of respective celltype programs in different topics such Cardiomycoyte in topic 25, Erythroblast in topic 9, Liver hepatoblasts topic 19</p>
</section>
<section id="16.-Computing-and-visualising-peaks-associated-with-each-topic">
<h2>16. Computing and visualising peaks associated with each topic<a class="headerlink" href="#16.-Computing-and-visualising-peaks-associated-with-each-topic" title="Link to this heading">#</a></h2>
<p>peaks associated with a topic should capture co-accesibility patterns</p>
<p>we visualise average accesibility of peaks (on a UMAP) in different celltypes and their association to a topic, to see if topics have captured co-accesibility patterns. Each point on the UMAP is a peak.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">umap_embedding_peaks</span><span class="p">,</span> <span class="n">topic_peak_embedding</span> <span class="o">=</span> <span class="n">compute_topic_peak_umap</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/data/saraswat/miniforge3/envs/scdori_test_env/lib/python3.10/site-packages/sklearn/utils/deprecation.py:151: FutureWarning: &#39;force_all_finite&#39; was renamed to &#39;ensure_all_finite&#39; in 1.6 and will be removed in 1.8.
  warnings.warn(
/data/saraswat/miniforge3/envs/scdori_test_env/lib/python3.10/site-packages/umap/umap_.py:1952: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(
INFO:scdori.downstream:Done. umap_embedding_peaks shape =&gt; (90000, 2) topic_embedding_peaks shape =&gt; (90000, 40)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## creating anndata with observations as peaks and values as topic association of each peak</span>
<span class="n">adata_peak</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">topic_peak_embedding</span><span class="p">)</span>
<span class="n">adata_peak</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Topic_&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">num_topics</span><span class="p">)]</span>
<span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
<span class="n">adata_peak</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">umap_embedding_peaks</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="n">atac_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># computing average accesiblity of peaks in each celltype</span>
<span class="n">atac_metacell</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">X</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">atac_metacell</span><span class="p">)</span>
<span class="n">aggregated_atac</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">atac_metacell</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mean&quot;</span><span class="p">])</span>
<span class="n">aggregated_atac</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">aggregated_atac</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">aggregated_atac</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">aggregated_atac</span><span class="p">)</span>

<span class="c1"># adding average accesibility of each peak in a celltype to peak anndata</span>
<span class="n">peak_celltype_df</span> <span class="o">=</span> <span class="n">aggregated_atac</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span><span class="o">.</span><span class="n">T</span>
<span class="n">peak_celltype_df</span> <span class="o">=</span> <span class="n">peak_celltype_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
<span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span><span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span><span class="n">peak_celltype_df</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_name</span> <span class="o">=</span> <span class="s1">&#39;Erythroid1&#39;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;Topic_9&quot;</span><span class="p">,</span><span class="s2">&quot;Topic_4&quot;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span> <span class="c1"># topic 9 and 4 are active in erythroids from heatmap above</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_71_0.png" src="../_images/notebooks_training_71_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_name</span> <span class="o">=</span> <span class="s1">&#39;NMP&#39;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;Topic_34&quot;</span><span class="p">,</span><span class="s2">&quot;Topic_7&quot;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_72_0.png" src="../_images/notebooks_training_72_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_name</span> <span class="o">=</span> <span class="s1">&#39;Spinal_cord&#39;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;Topic_7&quot;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_73_0.png" src="../_images/notebooks_training_73_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_name</span> <span class="o">=</span> <span class="s1">&#39;Cardiomyocytes&#39;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">celltype_name</span><span class="p">,</span><span class="s2">&quot;Topic_25&quot;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_74_0.png" src="../_images/notebooks_training_74_0.png" />
</div>
</div>
</section>
<section id="17.-Visualising-TF-binding-scores-on-peaks">
<h2>17. Visualising TF binding scores on peaks<a class="headerlink" href="#17.-Visualising-TF-binding-scores-on-peaks" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## adding insilico chipseq embeddings to peak anndata</span>


<span class="n">tf_names</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_type</span><span class="o">==</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># Create a dictionary for activator and repressor insilico-chipseq binding score</span>
<span class="n">tf_binding_data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">tf_name</span> <span class="o">+</span> <span class="s2">&quot;_activator_binding&quot;</span><span class="p">:</span> <span class="n">insilico_act</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf_name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_names</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">tf_binding_data</span><span class="o">.</span><span class="n">update</span><span class="p">({</span>
    <span class="n">tf_name</span> <span class="o">+</span> <span class="s2">&quot;_repressor_binding&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">insilico_rep</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">tf_name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tf_names</span><span class="p">)</span>
<span class="p">})</span>

<span class="c1"># Convert the dictionary to a DataFrame</span>
<span class="n">tf_binding_data_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tf_binding_data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

<span class="c1"># Concatenate new columns with existing obs</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">tf_binding_data_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">adata_peak</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span>  <span class="n">tf_binding_data_df</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#visualsing TF binding scores on peak umap</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">)</span>

<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">200</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Tbx5_activator_binding&#39;</span><span class="p">,</span><span class="s1">&#39;Gata1_activator_binding&#39;</span><span class="p">,</span><span class="s1">&#39;Ets1_activator_binding&#39;</span><span class="p">,</span><span class="s1">&#39;Sox2_activator_binding&#39;</span><span class="p">,</span><span class="s1">&#39;Hnf4a_activator_binding&#39;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Greens&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span><span class="n">vmax</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span><span class="n">sort_order</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_78_0.png" src="../_images/notebooks_training_78_0.png" />
</div>
</div>
</section>
</section>
<section id="Computing-eGRNs">
<h1>Computing eGRNs<a class="headerlink" href="#Computing-eGRNs" title="Link to this heading">#</a></h1>
<section id="18.-Computing-ATAC-based-GRNs-with-emprirical-significance">
<h2>18. Computing ATAC based GRNs with emprirical significance<a class="headerlink" href="#18.-Computing-ATAC-based-GRNs-with-emprirical-significance" title="Link to this heading">#</a></h2>
<p>these GRNs do not use evidence of TF-gene co-expression</p>
<p>activator GRNs here indicate if within a topic, peaks linked to a gene have accesible binding sites for a TF (from activator insilico-chipseq scores)</p>
<p>repressor GRNs here indicate if within a topic, peaks linked to a gene have non-accesible repressor binding sites for a TF (from repressor insilico-chipseq scores)</p>
<p>additionally we compute a background set of GRN values by shuffling insilico-chipseq scores, which are used to compute empirical significance</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="n">grn_act_atac</span> <span class="o">=</span> <span class="n">compute_atac_grn_activator_with_significance</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">cutoff_val</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">outdir</span><span class="o">=</span><span class="s2">&quot;grn_act_atac&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:Computing significant ATAC-derived TF–gene links for activators. Output =&gt; grn_act_atac
INFO:scdori.downstream:Processing Topic 1/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 1/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 1: 100%|██████████| 1000/1000 [00:10&lt;00:00, 92.32it/s]
INFO:scdori.downstream:Processing Topic 2/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 2/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 2: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.37it/s]
INFO:scdori.downstream:Processing Topic 3/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 3/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 3: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.42it/s]
INFO:scdori.downstream:Processing Topic 4/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 4/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 4: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.62it/s]
INFO:scdori.downstream:Processing Topic 5/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 5/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 5: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.92it/s]
INFO:scdori.downstream:Processing Topic 6/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 6/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 6: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.14it/s]
INFO:scdori.downstream:Processing Topic 7/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 7/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 7: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.48it/s]
INFO:scdori.downstream:Processing Topic 8/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 8/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 8: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.95it/s]
INFO:scdori.downstream:Processing Topic 9/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 9/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 9: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.85it/s]
INFO:scdori.downstream:Processing Topic 10/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 10/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 10: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.34it/s]
INFO:scdori.downstream:Processing Topic 11/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 11/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 11: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.93it/s]
INFO:scdori.downstream:Processing Topic 12/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 12/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 12: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.46it/s]
INFO:scdori.downstream:Processing Topic 13/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 13/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 13: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.79it/s]
INFO:scdori.downstream:Processing Topic 14/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 14/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 14: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.34it/s]
INFO:scdori.downstream:Processing Topic 15/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 15/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 15: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.48it/s]
INFO:scdori.downstream:Processing Topic 16/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 16/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 16: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.63it/s]
INFO:scdori.downstream:Processing Topic 17/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 17/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 17: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.71it/s]
INFO:scdori.downstream:Processing Topic 18/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 18/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 18: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.45it/s]
INFO:scdori.downstream:Processing Topic 19/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 19/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 19: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.08it/s]
INFO:scdori.downstream:Processing Topic 20/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 20/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 20: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.09it/s]
INFO:scdori.downstream:Processing Topic 21/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 21/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 21: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.39it/s]
INFO:scdori.downstream:Processing Topic 22/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 22/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 22: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.45it/s]
INFO:scdori.downstream:Processing Topic 23/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 23/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 23: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.27it/s]
INFO:scdori.downstream:Processing Topic 24/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 24/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 24: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.80it/s]
INFO:scdori.downstream:Processing Topic 25/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 25/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 25: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.66it/s]
INFO:scdori.downstream:Processing Topic 26/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 26/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 26: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.09it/s]
INFO:scdori.downstream:Processing Topic 27/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 27/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 27: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.47it/s]
INFO:scdori.downstream:Processing Topic 28/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 28/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 28: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.96it/s]
INFO:scdori.downstream:Processing Topic 29/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 29/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 29: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.67it/s]
INFO:scdori.downstream:Processing Topic 30/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 30/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 30: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.39it/s]
INFO:scdori.downstream:Processing Topic 31/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 31/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 31: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.28it/s]
INFO:scdori.downstream:Processing Topic 32/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 32/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 32: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.13it/s]
INFO:scdori.downstream:Processing Topic 33/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 33/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 33: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.69it/s]
INFO:scdori.downstream:Processing Topic 34/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 34/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 34: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.79it/s]
INFO:scdori.downstream:Processing Topic 35/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 35/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 35: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.62it/s]
INFO:scdori.downstream:Processing Topic 36/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 36/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 36: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.21it/s]
INFO:scdori.downstream:Processing Topic 37/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 37/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 37: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.31it/s]
INFO:scdori.downstream:Processing Topic 38/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 38/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 38: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.78it/s]
INFO:scdori.downstream:Processing Topic 39/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 39/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 39: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.76it/s]
INFO:scdori.downstream:Processing Topic 40/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 40/40
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 40: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.43it/s]
INFO:scdori.downstream:Completed computing activator ATAC GRNs.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[93]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># ATAC based GRN for repressors</span>
<span class="n">grn_rep_atac</span> <span class="o">=</span> <span class="n">compute_atac_grn_repressor_with_significance</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">cutoff_val</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">outdir</span><span class="o">=</span><span class="s2">&quot;grn_act_atac&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:Computing significant ATAC-derived TF–gene links for repressors. Output =&gt; grn_act_atac
INFO:scdori.downstream:Processing Topic 1/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 1/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 1: 100%|██████████| 1000/1000 [00:09&lt;00:00, 101.86it/s]
INFO:scdori.downstream:Processing Topic 2/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 2/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 2: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.26it/s]
INFO:scdori.downstream:Processing Topic 3/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 3/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 3: 100%|██████████| 1000/1000 [00:08&lt;00:00, 113.47it/s]
INFO:scdori.downstream:Processing Topic 4/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 4/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 4: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.23it/s]
INFO:scdori.downstream:Processing Topic 5/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 5/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 5: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.07it/s]
INFO:scdori.downstream:Processing Topic 6/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 6/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 6: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.09it/s]
INFO:scdori.downstream:Processing Topic 7/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 7/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 7: 100%|██████████| 1000/1000 [00:08&lt;00:00, 113.92it/s]
INFO:scdori.downstream:Processing Topic 8/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 8/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 8: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.17it/s]
INFO:scdori.downstream:Processing Topic 9/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 9/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 9: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.26it/s]
INFO:scdori.downstream:Processing Topic 10/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 10/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 10: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.98it/s]
INFO:scdori.downstream:Processing Topic 11/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 11/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 11: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.78it/s]
INFO:scdori.downstream:Processing Topic 12/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 12/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 12: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.73it/s]
INFO:scdori.downstream:Processing Topic 13/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 13/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 13: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.83it/s]
INFO:scdori.downstream:Processing Topic 14/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 14/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 14: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.58it/s]
INFO:scdori.downstream:Processing Topic 15/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 15/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 15: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.50it/s]
INFO:scdori.downstream:Processing Topic 16/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 16/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 16: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.45it/s]
INFO:scdori.downstream:Processing Topic 17/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 17/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 17: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.75it/s]
INFO:scdori.downstream:Processing Topic 18/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 18/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 18: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.75it/s]
INFO:scdori.downstream:Processing Topic 19/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 19/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 19: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.86it/s]
INFO:scdori.downstream:Processing Topic 20/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 20/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 20: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.38it/s]
INFO:scdori.downstream:Processing Topic 21/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 21/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 21: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.69it/s]
INFO:scdori.downstream:Processing Topic 22/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 22/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 22: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.21it/s]
INFO:scdori.downstream:Processing Topic 23/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 23/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 23: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.96it/s]
INFO:scdori.downstream:Processing Topic 24/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 24/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 24: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.71it/s]
INFO:scdori.downstream:Processing Topic 25/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 25/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 25: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.75it/s]
INFO:scdori.downstream:Processing Topic 26/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 26/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 26: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.05it/s]
INFO:scdori.downstream:Processing Topic 27/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 27/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 27: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.17it/s]
INFO:scdori.downstream:Processing Topic 28/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 28/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 28: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.04it/s]
INFO:scdori.downstream:Processing Topic 29/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 29/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 29: 100%|██████████| 1000/1000 [00:08&lt;00:00, 116.19it/s]
INFO:scdori.downstream:Processing Topic 30/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 30/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 30: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.97it/s]
INFO:scdori.downstream:Processing Topic 31/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 31/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 31: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.75it/s]
INFO:scdori.downstream:Processing Topic 32/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 32/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 32: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.56it/s]
INFO:scdori.downstream:Processing Topic 33/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 33/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 33: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.81it/s]
INFO:scdori.downstream:Processing Topic 34/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 34/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 34: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.83it/s]
INFO:scdori.downstream:Processing Topic 35/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 35/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 35: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.70it/s]
INFO:scdori.downstream:Processing Topic 36/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 36/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 36: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.16it/s]
INFO:scdori.downstream:Processing Topic 37/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 37/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 37: 100%|██████████| 1000/1000 [00:08&lt;00:00, 123.06it/s]
INFO:scdori.downstream:Processing Topic 38/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 38/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 38: 100%|██████████| 1000/1000 [00:08&lt;00:00, 122.79it/s]
INFO:scdori.downstream:Processing Topic 39/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 39/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 39: 100%|██████████| 1000/1000 [00:08&lt;00:00, 114.53it/s]
INFO:scdori.downstream:Processing Topic 40/40
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processing Topic 40/40
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Permutations for Topic 40: 100%|██████████| 1000/1000 [00:08&lt;00:00, 115.03it/s]
INFO:scdori.downstream:Completed computing repressor ATAC GRNs.
</pre></div></div>
</div>
</section>
<section id="19.-Computing-final-GRNs">
<h2>19. Computing final GRNs<a class="headerlink" href="#19.-Computing-final-GRNs" title="Link to this heading">#</a></h2>
<p>to compute these, we use the significant ATAC based GRNs derived previously and do element wise product with GRNs learnt by scDoRI incorproating TF - gene co-expression</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># calculating TF-expression per topic</span>
<span class="c1"># either from scdori model weights or from true data</span>
<span class="c1">#using true expression here</span>
<span class="n">tf_normalised</span> <span class="o">=</span> <span class="n">get_tf_expression</span><span class="p">(</span><span class="s2">&quot;True&quot;</span><span class="p">,</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">all_dataset_loader</span><span class="p">,</span><span class="n">rna_metacell</span><span class="p">,</span><span class="n">atac_metacell</span><span class="p">,</span><span class="n">num_cells</span><span class="p">,</span><span class="n">tf_indices</span><span class="p">,</span><span class="n">onehot_batch</span><span class="p">,</span><span class="n">config</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Extracting latent topics: 100%|██████████| 112/112 [01:27&lt;00:00,  1.28it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tf_normalised</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
torch.Size([40, 300])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># compute final GRNs which use the significant ATAC based GRNs derived above</span>
<span class="n">grn_act</span><span class="p">,</span> <span class="n">grn_rep</span> <span class="o">=</span> <span class="n">compute_significant_grn</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span> <span class="n">cutoff_val_activator</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span><span class="n">cutoff_val_repressor</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">tf_normalised</span><span class="o">=</span><span class="n">tf_normalised</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">(),</span> <span class="n">outdir</span><span class="o">=</span><span class="s2">&quot;grn_act_atac&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:Loading ATAC-derived GRNs...
INFO:scdori.downstream:Computing combined GRNs...
INFO:scdori.downstream:Saving computed GRNs...
INFO:scdori.downstream:GRN computation completed successfully.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># save regulons per TF</span>
<span class="n">save_regulons</span><span class="p">(</span><span class="n">grn_act</span><span class="p">,</span> <span class="n">tf_names</span><span class="o">=</span><span class="n">tf_names</span><span class="p">,</span> <span class="n">gene_names</span><span class="o">=</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">num_topics</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">num_topics</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;grn_act_atac&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;activator&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># save regulons per TF</span>
<span class="n">save_regulons</span><span class="p">(</span><span class="n">grn_rep</span><span class="p">,</span> <span class="n">tf_names</span><span class="o">=</span><span class="n">tf_names</span><span class="p">,</span> <span class="n">gene_names</span><span class="o">=</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">num_topics</span><span class="o">=</span><span class="n">model</span><span class="o">.</span><span class="n">num_topics</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;grn_act_atac&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;repressor&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># loading saved GRN</span>
<span class="n">grn_act</span> <span class="o">=</span>  <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;grn_act_atac/grn_activator__0.05.npy&#39;</span><span class="p">)</span>
<span class="n">grn_rep</span><span class="o">=</span>  <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;grn_act_atac/grn_repressor__0.05.npy&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">grn_act</span><span class="o">.</span><span class="n">shape</span> <span class="c1"># num_topics x num_tfs x num_genes</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(40, 300, 4000)
</pre></div></div>
</div>
</section>
<section id="20.-Computing-and-plotting-top-activator-TFs-per-topic">
<h2>20. Computing and plotting top activator TFs per topic<a class="headerlink" href="#20.-Computing-and-plotting-top-activator-TFs-per-topic" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="c1"># plotting TF activity across topics</span>
<span class="n">tf_names</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">gene_type</span><span class="o">==</span><span class="s1">&#39;TF&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>

<span class="c1"># plot top k activators per topic</span>
<span class="n">df_topic_activator</span><span class="p">,</span> <span class="n">top_regulators</span> <span class="o">=</span> <span class="n">get_top_activators_per_topic</span><span class="p">(</span>
    <span class="n">grn_act</span><span class="p">,</span>
    <span class="n">tf_names</span><span class="p">,</span>
    <span class="n">scdori_latent</span><span class="p">,</span>
    <span class="n">selected_topics</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">top_k</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">clamp_value</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">zscore</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span>
    <span class="n">out_fig</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Plotting top activator regulators per topic ===
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_92_1.png" src="../_images/notebooks_training_92_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Done plotting top regulators per topic ===
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_topic_activator</span>  <span class="c1"># matrix of Topic TF activities</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Alx1</th>
      <th>Ar</th>
      <th>Arid3c</th>
      <th>Arid5b</th>
      <th>Ascl1</th>
      <th>Ascl2</th>
      <th>Atf3</th>
      <th>Barx1</th>
      <th>Batf3</th>
      <th>Bcl11a</th>
      <th>...</th>
      <th>Zfp523</th>
      <th>Zfp69</th>
      <th>Zfp711</th>
      <th>Zic1</th>
      <th>Zic2</th>
      <th>Zic3</th>
      <th>Zic4</th>
      <th>Zic5</th>
      <th>Zscan10</th>
      <th>Zscan4c</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Topic_0</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.113955</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>0.164900</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_1</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>2.412600</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>6.166309</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_3</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>1.709113</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>0.283313</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_5</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.479874</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>0.624574</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>1.990908</td>
      <td>0.903295</td>
      <td>0.0</td>
      <td>0.419149</td>
      <td>1.018452</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_6</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.044312</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>3.561999</td>
      <td>0.0</td>
      <td>0.685740</td>
      <td>0.145646</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_7</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.598857</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_8</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.529142</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>2.522007</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>0.289952</td>
      <td>0.114372</td>
      <td>0.0</td>
      <td>4.639709</td>
      <td>1.842462</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_9</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_10</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.373513</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.102275</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_11</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_12</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.072807</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>0.112615</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_13</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>5.366496</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>2.602772</td>
      <td>2.711501</td>
      <td>0.0</td>
      <td>3.513520</td>
      <td>5.409577</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_14</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.406806</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.023502</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.217899</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_15</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.282287</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>0.024857</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_16</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.331571</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_17</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.150101</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>0.050481</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.305153</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_18</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>0.345635</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>2.662320</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>0.222181</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_19</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>1.664095</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_20</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.325911</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_21</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>2.591520</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.177927</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.283884</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_22</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.562149</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.147956</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_23</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>0.051448</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>0.108123</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>0.153288</td>
      <td>0.652382</td>
      <td>0.0</td>
      <td>-0.158550</td>
      <td>0.971483</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_24</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.263331</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>0.463549</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.158753</td>
      <td>0.0</td>
      <td>0.764304</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>0.097052</td>
      <td>0.169522</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_25</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.677782</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_26</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.667436</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>6.166433</td>
      <td>6.115410</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_27</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>2.092698</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_28</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>1.328635</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_29</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>6.166414</td>
      <td>2.135152</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.028738</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_30</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.248851</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>3.723565</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>0.235461</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_31</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>0.015134</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.298730</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.129481</td>
      <td>-0.348830</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_32</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.270752</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_33</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_34</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_35</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>0.734621</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_36</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>0.048776</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>0.080043</td>
      <td>0.0</td>
      <td>-0.237916</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>0.683270</td>
      <td>0.336786</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_37</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>2.345082</td>
      <td>0.0</td>
      <td>6.166427</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>4.818160</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>-0.365167</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>-0.348377</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_38</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.634754</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>-0.286531</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>4.788070</td>
      <td>-0.431009</td>
      <td>0.0</td>
      <td>0.547308</td>
      <td>-0.370249</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>Topic_39</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.689207</td>
      <td>0.0</td>
      <td>-0.158113</td>
      <td>-0.158110</td>
      <td>0.0</td>
      <td>-0.224782</td>
      <td>0.657307</td>
      <td>...</td>
      <td>0.0</td>
      <td>-0.158114</td>
      <td>-0.222186</td>
      <td>0.0</td>
      <td>0.094880</td>
      <td>0.691460</td>
      <td>0.0</td>
      <td>0.153598</td>
      <td>0.430678</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>40 rows × 300 columns</p>
</div></div>
</div>
</section>
<section id="21.-Computing-and-plotting-TF-activity-per-cell">
<h2>21. Computing and plotting TF activity per cell<a class="headerlink" href="#21.-Computing-and-plotting-TF-activity-per-cell" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># computing TF activity per cell</span>
<span class="n">cell_tf_act</span> <span class="o">=</span> <span class="n">compute_activator_tf_activity_per_cell</span><span class="p">(</span>
    <span class="n">grn_act</span><span class="p">,</span>
    <span class="n">tf_names</span><span class="p">,</span>
    <span class="n">scdori_latent</span><span class="p">,</span>
    <span class="n">selected_topics</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">clamp_value</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">zscore</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Computing TF activity per cell ===
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># aggregating activity per celltype</span>
<span class="n">df_celltype_tf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_tf_act</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">tf_names</span><span class="p">)</span>
<span class="n">df_celltype_tf</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">df_celltype_tf</span><span class="o">=</span><span class="n">df_celltype_tf</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;celltype&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">df_celltype_tf</span><span class="o">=</span><span class="n">df_celltype_tf</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">df_celltype_tf</span> <span class="o">=</span> <span class="n">df_celltype_tf</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">(</span><span class="n">df_celltype_tf</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)]</span> <span class="c1"># removing TF with 0/Nan activity</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_220372/1244025385.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_celltype_tf=df_celltype_tf.groupby(&#39;celltype&#39;).mean()
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># top TFs per celltype</span>
<span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_celltype_tf</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">df_celltype_tf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:</span><span class="mi">5</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Msx1      2.812401
Pitx1     2.482060
Prdm6     2.420417
Tbx4      2.246442
Plagl1    2.209494
Name: Allantois, dtype: float32
Foxa2     3.318382
Sp5       3.230514
Prdm1     3.220130
Lhx1      2.549573
Tcf7l2    2.466181
Name: Anterior_Primitive_Streak, dtype: float32
Thap1     3.457717
Myb       3.403746
Fli1      3.332256
Nfatc1    3.291261
Hhex      3.053652
Name: Blood_progenitors_1, dtype: float32
Myb       3.962984
Thap1     3.834107
Fli1      3.445354
Runx1     3.366277
Nfatc1    3.314607
Name: Blood_progenitors_2, dtype: float32
Nkx2-5    4.915843
Tbx2      4.889453
Irx4      4.889432
Tbx5      4.846557
Esrrg     4.772815
Name: Cardiomyocytes, dtype: float32
Cdx2     2.382280
Hoxb9    2.251371
Hoxb8    2.235967
Hoxa7    2.233847
Hoxb4    2.232038
Name: Caudal_Mesoderm, dtype: float32
Cux2       2.315070
Bcl11a     2.111253
Zic3       2.101088
Pou5f1     2.095708
Zscan10    2.075776
Name: Caudal_epiblast, dtype: float32
Cux2       2.459110
Zscan10    2.317297
Bcl11a     2.295461
Zic2       2.286987
Pou5f1     2.271102
Name: Caudal_neurectoderm, dtype: float32
Foxa2     3.013058
Prdm1     2.905914
Sp5       2.884930
Sox17     2.415743
Tcf7l2    2.328682
Name: Def._endoderm, dtype: float32
Irf2     5.552412
Elk3     5.518743
Zfp69    5.487562
Erg      5.487561
Nr5a2    5.487561
Name: Endothelium, dtype: float32
Zic5       3.557174
Bcl11b     3.491947
Bcl11a     3.369754
Pou5f1     3.333444
Zscan10    3.289199
Name: Epiblast, dtype: float32
Rreb1    3.154777
Gfi1b    3.113449
Runx1    3.113264
E2f2     3.106583
Gata1    3.080394
Name: Erythroid1, dtype: float32
Nfe2     3.708780
Klf1     3.708776
Foxo3    3.698624
Jun      3.691791
Gata1    3.602191
Name: Erythroid2, dtype: float32
Nfe2     4.405540
Klf1     4.405535
Jun      4.351606
Foxo3    4.351569
Gata1    4.113173
Name: Erythroid3, dtype: float32
Elf5      3.741948
Zfp382    3.741423
Nr4a1     3.741415
Npas2     3.741409
Esrrb     3.737677
Name: ExE_ectoderm, dtype: float32
Maf       2.658887
Nr3c2     2.658283
Esx1      2.657149
Foxa3     2.657129
Nfatc2    2.657127
Name: ExE_endoderm, dtype: float32
Osr1     2.054457
Hoxc9    2.054455
Hoxd4    1.987622
Hoxc6    1.947457
Hoxa7    1.909860
Name: ExE_mesoderm, dtype: float32
Otx1    2.527822
Pax2    2.510998
Pax8    2.489977
Pax5    2.481016
En1     2.467573
Name: Forebrain_Midbrain_Hindbrain, dtype: float32
Prdm1     2.518131
Sox17     2.458320
Foxa2     2.443587
Tcf7l2    2.397369
Sp5       2.148218
Name: Gut, dtype: float32
Irf2      3.099161
Hhex      3.060217
Zfp711    3.056182
Elk3      3.024903
Zfp69     2.948663
Name: Haematoendothelial_progenitors, dtype: float32
Osr1     1.688312
Hoxc9    1.688310
Hoxb1    1.664535
Hes7     1.655115
Hoxd4    1.594885
Name: Intermediate_mesoderm, dtype: float32
Creb3l1    2.819529
Myrf       2.815305
Plagl1     2.809360
Pitx2      2.791425
Pitx1      2.728289
Name: Mesenchyme, dtype: float32
Lhx1     2.732252
Sp5      1.933315
Foxa2    1.605808
Prdm1    1.563257
Mecom    1.357563
Name: Mixed_mesoderm, dtype: float32
Hoxb9    2.574918
Hoxc4    2.534038
Hoxb4    2.478113
Hoxa3    2.455356
Hoxa9    2.437630
Name: NMP, dtype: float32
Lhx1    3.333681
Sp5     2.254464
T       2.248332
Lef1    2.052773
Tbx6    2.003616
Name: Nascent_mesoderm, dtype: float32
Tfap2b    5.577940
Foxd3     5.577939
Sox10     5.577938
Ets1      5.574512
Sox9      5.568067
Name: Neural_crest, dtype: float32
Sp5      3.020447
Foxa2    2.908019
Prdm1    2.750876
Egr1     2.424358
Sox17    2.161388
Name: Notochord, dtype: float32
Zbtb2    1.922414
Zic2     1.731415
Pax7     1.538206
Lmx1a    1.538201
Cdx2     1.494526
Name: PGC, dtype: float32
Foxc2    3.293137
Tbx18    3.293136
Meox1    3.292907
Six1     3.227916
Foxp2    3.137960
Name: Paraxial_mesoderm, dtype: float32
Sox7     6.944213
Jdp2     6.408046
Snai1    6.376996
Stat1    6.172889
Klf5     6.024989
Name: Parietal_endoderm, dtype: float32
Meis1    2.291860
Gata6    1.885065
Nr2f2    1.827772
Esrrg    1.799770
Mef2c    1.728161
Name: Pharyngeal_mesoderm, dtype: float32
Zic3       2.615363
Pou5f1     2.614649
Zscan10    2.540146
Bcl11a     2.537390
Cux2       2.510679
Name: Primitive_Streak, dtype: float32
Zic5       2.991289
Bcl11b     2.979502
Bcl11a     2.973205
Zscan10    2.922971
Pou5f1     2.922059
Name: Rostral_neurectoderm, dtype: float32
Hoxb1    3.046862
Hes7     2.999344
T        2.960349
Tbx6     2.700277
Lef1     2.546784
Name: Somitic_mesoderm, dtype: float32
Nkx6-1    3.304471
Pax6      3.275482
Foxb1     3.182235
Rarb      3.176786
Hes5      3.051524
Name: Spinal_cord, dtype: float32
Trp63     3.160946
Grhl2     3.118778
Batf3     2.599621
Tfap2a    2.041930
Gata3     1.862946
Name: Surface_ectoderm, dtype: float32
Sox17     3.455087
Foxa2     3.350312
Tcf7l2    3.288827
Prdm1     3.182143
Sp5       2.886621
Name: Visceral_endoderm, dtype: float32
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">df_celltype_tf</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">4</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">15</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.matrix.ClusterGrid at 0x7f6ecd083190&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_98_1.png" src="../_images/notebooks_training_98_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># visualing TF activity on UMAP</span>
<span class="n">df_cell_tf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_tf_act</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="n">s</span><span class="o">+</span><span class="s1">&#39;_activity&#39;</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">tf_names</span><span class="p">])</span>
<span class="n">df_cell_tf</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
<span class="n">obs_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span><span class="n">df_cell_tf</span><span class="p">],</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">df_cell_tf</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">=</span><span class="n">df_cell_tf</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span><span class="s1">&#39;Gata1_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Sox10_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Gata6_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Ets1_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Hnf4a_activity&#39;</span><span class="p">,</span><span class="s1">&#39;T_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Prrx1_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Tbx5_activity&#39;</span><span class="p">,</span><span class="s1">&#39;Pou5f1_activity&#39;</span><span class="p">],</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">2</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_101_0.png" src="../_images/notebooks_training_101_0.png" />
</div>
</div>
</section>
<section id="22.-visualising-downstream-target-genes-of-a-TF">
<h2>22. visualising downstream target genes of a TF<a class="headerlink" href="#22.-visualising-downstream-target-genes-of-a-TF" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[83]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tf_plot</span><span class="o">=</span><span class="s2">&quot;Gata4&quot;</span>

<span class="n">tf_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">tf_names</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">tf_plot</span><span class="p">)</span>
<span class="n">tf_index</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[83]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
68
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[94]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_topic_activator</span><span class="p">[</span><span class="n">tf_plot</span><span class="p">]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[94]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Topic_3     3.130074
Topic_29    2.430753
Topic_19    1.713511
Topic_39    1.622072
Topic_25    1.532704
Name: Gata4, dtype: float32
</pre></div></div>
</div>
<p>since Gata4 is active in multiple topics ( and associated cellltype), we obtain different targets for it in the respective context</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[112]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">topic_num</span> <span class="o">=</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">19</span><span class="p">]</span> <span class="c1"># endodermal topics</span>
<span class="n">target_gene_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">grn_act</span><span class="p">[</span><span class="n">topic_num</span><span class="p">,</span><span class="n">tf_index</span><span class="p">,:]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span><span class="mf">0.0</span> <span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># adjust this value to get more stringent/ strongly regulated downstream taregts</span>
<span class="n">genes_endoderm</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">target_gene_idx</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[126]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes_endoderm</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[126]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;Abcc2&#39;, &#39;Abcc4&#39;, &#39;Abcc6&#39;, &#39;Ablim1&#39;, &#39;Adcy9&#39;, &#39;Aff2&#39;, &#39;Agbl4&#39;, &#39;Agtr1b&#39;,
       &#39;Akr1d1&#39;, &#39;Alas1&#39;,
       ...
       &#39;Olig1&#39;, &#39;Prdm6&#39;, &#39;Rfx6&#39;, &#39;Snai3&#39;, &#39;Sox17&#39;, &#39;Stat2&#39;, &#39;Stat3&#39;, &#39;Tcf7l2&#39;,
       &#39;Tead4&#39;, &#39;Zfp523&#39;],
      dtype=&#39;object&#39;, length=387)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[114]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">topic_num</span> <span class="o">=</span> <span class="p">[</span><span class="mi">25</span><span class="p">]</span>  <span class="c1"># cardiomyocyte specific topic</span>
<span class="n">target_gene_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">grn_act</span><span class="p">[</span><span class="n">topic_num</span><span class="p">,</span><span class="n">tf_index</span><span class="p">,:]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span><span class="mf">0.0</span> <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">genes_cardiomyocytes</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">target_gene_idx</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[127]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes_cardiomyocytes</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[127]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Index([&#39;Ablim1&#39;, &#39;Acta1&#39;, &#39;Aff2&#39;, &#39;Agbl4&#39;, &#39;Ankrd1&#39;, &#39;Arhgap30&#39;, &#39;Arl4d&#39;,
       &#39;Bmper&#39;, &#39;Cacna1i&#39;, &#39;Camk2d&#39;, &#39;Casz1&#39;, &#39;Ccdc141&#39;, &#39;Cdh3&#39;, &#39;Cdkn1c&#39;,
       &#39;Cgnl1&#39;, &#39;Dag1&#39;, &#39;Dsg3&#39;, &#39;Dsp&#39;, &#39;Epb41l3&#39;, &#39;Erich6b&#39;, &#39;Fam151a&#39;,
       &#39;Frmd4b&#39;, &#39;Ghr&#39;, &#39;Gpx3&#39;, &#39;Grb14&#39;, &#39;Grin2c&#39;, &#39;Grin3a&#39;, &#39;Hcn4&#39;, &#39;Hopx&#39;,
       &#39;Igf2&#39;, &#39;Il1r1&#39;, &#39;Il1rl1&#39;, &#39;Ins2&#39;, &#39;Kif19a&#39;, &#39;Lama1&#39;, &#39;Lilr4b&#39;, &#39;Lmo7&#39;,
       &#39;Lrrc43&#39;, &#39;Ly75&#39;, &#39;Meikin&#39;, &#39;Mical2&#39;, &#39;Mmp14&#39;, &#39;Mpz&#39;, &#39;Mttp&#39;, &#39;Myh10&#39;,
       &#39;Myh6&#39;, &#39;Nebl&#39;, &#39;Nedd4l&#39;, &#39;Nexn&#39;, &#39;Peg3&#39;, &#39;Plac8&#39;, &#39;Popdc2&#39;, &#39;Popdc3&#39;,
       &#39;Ptk2b&#39;, &#39;Rcan2&#39;, &#39;Rcsd1&#39;, &#39;Rec114&#39;, &#39;Rhoh&#39;, &#39;Rspo1&#39;, &#39;Sipa1l2&#39;,
       &#39;Slc16a12&#39;, &#39;Slc16a5&#39;, &#39;Smarcd3&#39;, &#39;Svep1&#39;, &#39;Tfpi&#39;, &#39;Tiparp&#39;, &#39;Tln2&#39;,
       &#39;Tmem202&#39;, &#39;Treml1&#39;, &#39;Trpc6&#39;, &#39;Trpm5&#39;, &#39;Ttc22&#39;, &#39;Unc5b&#39;, &#39;Vav3&#39;, &#39;Dpf3&#39;,
       &#39;Gata4&#39;, &#39;Prdm6&#39;, &#39;Rfx6&#39;],
      dtype=&#39;object&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[119]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plotting downstream expression of target genes</span>
<span class="n">rna_metacell</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]</span>
<span class="c1"># normalsing raw counts</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span><span class="n">genes_endoderm</span><span class="p">,</span> <span class="n">score_name</span><span class="o">=</span><span class="s2">&quot;Gata4_endoderm_target&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span><span class="n">genes_cardiomyocytes</span><span class="p">,</span> <span class="n">score_name</span><span class="o">=</span><span class="s2">&quot;Gata4_cardiomyocyte_target&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: adata.X seems to be already log-transformed.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[139]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="c1"># visualing cell-types on scDoRI computed UMAP</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">200</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span><span class="s1">&#39;Gata4&#39;</span><span class="p">,</span><span class="s2">&quot;Gata4_endoderm_target&quot;</span><span class="p">,</span><span class="s2">&quot;Gata4_cardiomyocyte_target&quot;</span><span class="p">],</span><span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">outline_color</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="s1">&#39;black&#39;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">,</span><span class="n">legend_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_111_0.png" src="../_images/notebooks_training_111_0.png" />
</div>
</div>
<p>we can clearly see that scDoRI finds differential downstream targets for the same TF in different contexts</p>
<p>we can visualise the TF binding profiles to confirm that these differences are coming from chromatin differences between states</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[138]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="c1">#visualsing TF binding scores and topic values where TF is active, on peak umap</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># topic 3 and 19 are endoderm related and topic 25 is cardiomyocyte specific</span>
<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">150</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Gata4_activator_binding&#39;</span><span class="p">]</span> <span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Greens&#39;</span><span class="p">,</span> <span class="n">sort_order</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_113_0.png" src="../_images/notebooks_training_113_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[135]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="c1">#visualsing TF binding scores and topic values where TF is active, on peak umap</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># topic 3 and 19 are endoderm related and topic 25 is cardiomyocyte specific</span>
<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">200</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_peak</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Topic_3&#39;</span><span class="p">,</span><span class="s1">&#39;Topic_19&#39;</span><span class="p">,</span><span class="s1">&#39;Topic_25&#39;</span><span class="p">]</span> <span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Greens&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort_order</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_114_0.png" src="../_images/notebooks_training_114_0.png" />
</div>
</div>
<p>we can see that Gata4 binds to regulatory regions associated with different topics and can regulate different set of genes in those topics respectively</p>
</section>
<section id="23.-Repressor-analysis">
<h2>23. Repressor analysis<a class="headerlink" href="#23.-Repressor-analysis" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[140]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_topic_repressor</span><span class="p">,</span> <span class="n">top_regulators_repressor</span> <span class="o">=</span> <span class="n">get_top_repressor_per_topic</span><span class="p">(</span>
    <span class="n">grn_rep</span><span class="p">,</span>
    <span class="n">tf_names</span><span class="p">,</span>
    <span class="n">scdori_latent</span><span class="p">,</span>
    <span class="n">selected_topics</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">top_k</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">clamp_value</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">zscore</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span>
    <span class="n">out_fig</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Plotting top repressor regulators per topic ===
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_117_1.png" src="../_images/notebooks_training_117_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Done plotting top repressor regulators per topic ===
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[141]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_tf_rep</span> <span class="o">=</span> <span class="n">compute_repressor_tf_activity_per_cell</span><span class="p">(</span>
    <span class="n">grn_rep</span><span class="p">,</span>
    <span class="n">tf_names</span><span class="p">,</span>
    <span class="n">scdori_latent</span><span class="p">,</span>
    <span class="n">selected_topics</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">clamp_value</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">,</span>
    <span class="n">zscore</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:scdori.downstream:=== Computing TF activity per cell ===
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[142]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_celltype_tf_rep</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_tf_rep</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">tf_names</span><span class="p">)</span>
<span class="n">df_celltype_tf_rep</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">df_celltype_tf_rep</span><span class="o">=</span><span class="n">df_celltype_tf_rep</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;celltype&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_220372/3645810683.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_celltype_tf_rep=df_celltype_tf_rep.groupby(&#39;celltype&#39;).mean()
</pre></div></div>
</div>
</section>
<section id="24.-visualising-enhancer-gene-links">
<h2>24. visualising enhancer gene links<a class="headerlink" href="#24.-visualising-enhancer-gene-links" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[353]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># peaks gene links used by scdori</span>
<span class="n">gene_peak</span> <span class="o">=</span> <span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">gene_peak_factor_learnt</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span><span class="o">*</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">gene_peak_factor_fixed</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[356]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_peak</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[356]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(4000, 90000)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_name</span> <span class="o">=</span><span class="s2">&quot;Tal1&quot;</span>
<span class="n">gene_index</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>

<span class="n">enhancers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">gene_peak</span><span class="p">[</span><span class="n">gene_index</span><span class="p">,:]</span> <span class="o">&gt;</span> <span class="mf">0.99</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># change this threshold to obtain more links</span>
<span class="n">enhancers</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">enhancers</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>plotting accesibility of Tal1 enhancers across celltypes</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[425]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">peak_gene_celltype</span> <span class="o">=</span> <span class="n">peak_celltype_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enhancers</span><span class="p">]</span>
<span class="n">peak_gene_celltype</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[425]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Allantois</th>
      <th>Anterior_Primitive_Streak</th>
      <th>Blood_progenitors_1</th>
      <th>Blood_progenitors_2</th>
      <th>Cardiomyocytes</th>
      <th>Caudal_Mesoderm</th>
      <th>Caudal_epiblast</th>
      <th>Caudal_neurectoderm</th>
      <th>Def._endoderm</th>
      <th>Endothelium</th>
      <th>...</th>
      <th>PGC</th>
      <th>Paraxial_mesoderm</th>
      <th>Parietal_endoderm</th>
      <th>Pharyngeal_mesoderm</th>
      <th>Primitive_Streak</th>
      <th>Rostral_neurectoderm</th>
      <th>Somitic_mesoderm</th>
      <th>Spinal_cord</th>
      <th>Surface_ectoderm</th>
      <th>Visceral_endoderm</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>chr4:115049734-115050334</th>
      <td>-0.471083</td>
      <td>-0.587658</td>
      <td>2.084388</td>
      <td>1.262422</td>
      <td>-0.374860</td>
      <td>-0.283599</td>
      <td>-0.426946</td>
      <td>0.095904</td>
      <td>-0.348895</td>
      <td>3.098126</td>
      <td>...</td>
      <td>-0.587658</td>
      <td>-0.396268</td>
      <td>-0.571087</td>
      <td>-0.121929</td>
      <td>-0.418987</td>
      <td>-0.408271</td>
      <td>-0.475751</td>
      <td>-0.442686</td>
      <td>-0.477456</td>
      <td>-0.311563</td>
    </tr>
    <tr>
      <th>chr4:115071585-115072185</th>
      <td>-0.484199</td>
      <td>-0.387654</td>
      <td>1.670970</td>
      <td>1.601316</td>
      <td>-0.436656</td>
      <td>-0.601173</td>
      <td>-0.345486</td>
      <td>-0.710103</td>
      <td>-0.483152</td>
      <td>3.579220</td>
      <td>...</td>
      <td>-0.710103</td>
      <td>-0.387557</td>
      <td>-0.673809</td>
      <td>-0.424735</td>
      <td>-0.382381</td>
      <td>-0.373902</td>
      <td>-0.579387</td>
      <td>-0.430436</td>
      <td>-0.523610</td>
      <td>-0.559376</td>
    </tr>
    <tr>
      <th>chr4:114677387-114677987</th>
      <td>2.833133</td>
      <td>-0.809878</td>
      <td>-0.706228</td>
      <td>0.462337</td>
      <td>-0.790806</td>
      <td>0.652484</td>
      <td>0.088338</td>
      <td>-0.926896</td>
      <td>0.048584</td>
      <td>-0.771792</td>
      <td>...</td>
      <td>-0.384917</td>
      <td>-0.183126</td>
      <td>-0.896655</td>
      <td>-0.327314</td>
      <td>-0.495454</td>
      <td>-0.587162</td>
      <td>0.002362</td>
      <td>-0.556499</td>
      <td>0.311325</td>
      <td>-0.110331</td>
    </tr>
    <tr>
      <th>chr4:115070122-115070722</th>
      <td>-0.418941</td>
      <td>0.005167</td>
      <td>1.908539</td>
      <td>1.283993</td>
      <td>-0.435678</td>
      <td>-0.610410</td>
      <td>-0.464752</td>
      <td>-0.610410</td>
      <td>-0.485386</td>
      <td>0.334314</td>
      <td>...</td>
      <td>-0.610410</td>
      <td>-0.441879</td>
      <td>-0.496640</td>
      <td>-0.450805</td>
      <td>-0.319608</td>
      <td>-0.264801</td>
      <td>-0.483289</td>
      <td>-0.422015</td>
      <td>-0.483358</td>
      <td>-0.399455</td>
    </tr>
  </tbody>
</table>
<p>4 rows × 37 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[426]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">peak_celltype_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enhancers</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">5</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[426]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;seaborn.matrix.ClusterGrid at 0x7f6f66493160&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_126_1.png" src="../_images/notebooks_training_126_1.png" />
</div>
</div>
<p>plotting Tal1 expression and net accesibility of its predcited enhancers across celltypes</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[410]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atac_metacell</span><span class="o">.</span><span class="n">X</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;counts&#39;</span><span class="p">]</span>
<span class="c1"># normalsing raw counts</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">atac_metacell</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes</span><span class="p">(</span><span class="n">atac_metacell</span><span class="p">,</span> <span class="n">enhancers</span><span class="p">,</span> <span class="n">score_name</span><span class="o">=</span><span class="s1">&#39;Tal1_enhancer_accesibility&#39;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[428]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Tal1_enhancer_accesibility&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">atac_metacell</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Tal1_enhancer_accesibility&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[433]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s2">&quot;whitegrid&quot;</span><span class="p">)</span>
<span class="k">with</span> <span class="n">plt</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s2">&quot;figure.figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="s2">&quot;figure.dpi&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">200</span><span class="p">)}):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">rna_metacell</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span><span class="s1">&#39;Tal1&#39;</span><span class="p">,</span><span class="s2">&quot;Tal1_enhancer_accesibility&quot;</span><span class="p">],</span><span class="n">add_outline</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">outline_color</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="s1">&#39;black&#39;</span><span class="p">),</span><span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;YlGnBu&#39;</span><span class="p">,</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">legend_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_training_130_0.png" src="../_images/notebooks_training_130_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Loading and preparing data for training and model initialisation</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#1.-load-data">1. load data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#2.-computing-indices-of-genes-which-are-TFs-and-setting-number-of-cells-per-metacell-(-set-to-1-for-single-cell-data)">2. computing indices of genes which are TFs and setting number of cells per metacell ( set to 1 for single cell data)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#3.-onehot-encoding-the-batch-column-for-entire-dataset">3. onehot encoding the batch column for entire dataset</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#4.-making-train-and-evaluation-datasets">4. making train and evaluation datasets</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#5.-Build-scDoRI-model-using-parameters-from-config-file">5. Build scDoRI model using parameters from config file</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#6.-initialising-scDoRI-model-with-precomputed-matrices-and-setting-gradients">6. initialising scDoRI model with precomputed matrices and setting gradients</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#Train-Phase-1-of-scDoRI-model">Train Phase 1 of scDoRI model</a></li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#Train-Phase-2-of-scDoRI-model">Train Phase 2 of scDoRI model</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#7.-Load-best-checkpoint-from-Phase-1">7. Load best checkpoint from Phase 1</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#8.-Set-gradients-for-Phase-2-training">8. Set gradients for Phase 2 training</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#9.-Phase-2-training-and-saving-model-weights">9. Phase 2 training and saving model weights</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#Downstream-analysis">Downstream analysis</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#10.-Load-best-checkpoint-model">10. Load best checkpoint model</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#11.-Computing-and-visualising-latent-topic-activity-per-cell">11. Computing and visualising latent topic activity per cell</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#12.-Computing-average-topic-activation-in-different-celltypes/groups">12. Computing average topic activation in different celltypes/groups</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#13.-single-cell-level-visualisation-of-scDoRI-Topics">13. single cell level visualisation of scDoRI Topics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#14.-Computing-Top-genes-per-topic">14. Computing Top genes per topic</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#15.-Performing-gene-set-enrichment-analysis-on-each-topic">15. Performing gene-set enrichment analysis on each topic</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#16.-Computing-and-visualising-peaks-associated-with-each-topic">16. Computing and visualising peaks associated with each topic</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#17.-Visualising-TF-binding-scores-on-peaks">17. Visualising TF binding scores on peaks</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#Computing-eGRNs">Computing eGRNs</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#18.-Computing-ATAC-based-GRNs-with-emprirical-significance">18. Computing ATAC based GRNs with emprirical significance</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#19.-Computing-final-GRNs">19. Computing final GRNs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#20.-Computing-and-plotting-top-activator-TFs-per-topic">20. Computing and plotting top activator TFs per topic</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#21.-Computing-and-plotting-TF-activity-per-cell">21. Computing and plotting TF activity per cell</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#22.-visualising-downstream-target-genes-of-a-TF">22. visualising downstream target genes of a TF</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#23.-Repressor-analysis">23. Repressor analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#24.-visualising-enhancer-gene-links">24. visualising enhancer gene links</a></li>
</ul>
</li>
</ul>

  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/notebooks/training.ipynb.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Manu Saraswat.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>