<!DOCTYPE html>

<html lang="en" data-content_root="../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>preprocessing_pipeline.filtering &#8212; scDoRI  documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css?v=27fed22d" />
    <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for preprocessing_pipeline.filtering</h1><div class="highlight"><pre>
<span></span><span class="c1"># filtering.py</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="intersect_cells">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.filtering.intersect_cells">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">intersect_cells</span><span class="p">(</span>
    <span class="n">data_rna</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">data_atac</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Keep only the cells that are present in both RNA and ATAC datasets.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_rna : anndata.AnnData</span>
<span class="sd">        The RNA single-cell data.</span>
<span class="sd">    data_atac : anndata.AnnData</span>
<span class="sd">        The ATAC single-cell data.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    (data_rna_sub, data_atac_sub) : tuple of anndata.AnnData</span>
<span class="sd">        Two AnnData objects that share the same set of cells, in the same order.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The function finds the intersection of `obs_names` (cell barcodes) between</span>
<span class="sd">      the two AnnData objects.</span>
<span class="sd">    - It logs the new shapes of the intersected RNA and ATAC data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">common_idx</span> <span class="o">=</span> <span class="n">data_rna</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">data_atac</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
    <span class="n">data_rna_sub</span> <span class="o">=</span> <span class="n">data_rna</span><span class="p">[</span><span class="n">common_idx</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">data_atac_sub</span> <span class="o">=</span> <span class="n">data_atac</span><span class="p">[</span><span class="n">common_idx</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Intersected cells: now RNA=</span><span class="si">{</span><span class="n">data_rna_sub</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">, ATAC=</span><span class="si">{</span><span class="n">data_atac_sub</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data_rna_sub</span><span class="p">,</span> <span class="n">data_atac_sub</span></div>


<div class="viewcode-block" id="remove_mitochondrial_genes">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.filtering.remove_mitochondrial_genes">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">remove_mitochondrial_genes</span><span class="p">(</span>
    <span class="n">data_rna</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">mito_prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;mt-&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove mitochondrial genes from the RNA data based on a gene name prefix.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_rna : anndata.AnnData</span>
<span class="sd">        The RNA AnnData containing gene expression counts.</span>
<span class="sd">    mito_prefix : str, optional</span>
<span class="sd">        The prefix used to identify mitochondrial genes. Default is &quot;mt-&quot;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    anndata.AnnData</span>
<span class="sd">        A new AnnData object without mitochondrial genes. The original data</span>
<span class="sd">        is not modified in-place.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Mitochondrial genes are identified by checking if gene names start with</span>
<span class="sd">      `mito_prefix` (case-insensitive).</span>
<span class="sd">    - A boolean column &quot;mt&quot; is added to `data_rna.var` to indicate whether</span>
<span class="sd">      each gene was marked as mitochondrial. This subset is then removed</span>
<span class="sd">      from the data.</span>
<span class="sd">    - Logs the number of removed genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_rna</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">mito_prefix</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">data_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
    <span class="n">keep_mask</span> <span class="o">=</span> <span class="o">~</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">]</span>
    <span class="n">data_rna_sub</span> <span class="o">=</span> <span class="n">data_rna</span><span class="p">[:,</span> <span class="n">keep_mask</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">dropped</span> <span class="o">=</span> <span class="n">data_rna</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">data_rna_sub</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Removed </span><span class="si">{</span><span class="n">dropped</span><span class="si">}</span><span class="s2"> mitochondrial genes with prefix=</span><span class="si">{</span><span class="n">mito_prefix</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data_rna_sub</span></div>

</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">scDoRI</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../preprocessing_pipeline.html">Preprocessing Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../scdori.html">scDORI training</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks.html">Notebook Examples</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Manu Saraswat.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
    </div>

    

    
  </body>
</html>