<!DOCTYPE html>

<html lang="en" data-content_root="../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>preprocessing_pipeline.gene_selection &#8212; scDoRI  documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../../_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css?v=27fed22d" />
    <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for preprocessing_pipeline.gene_selection</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="c1">#import muon as mu</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">gtfparse</span><span class="w"> </span><span class="kn">import</span> <span class="n">read_gtf</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="load_gtf">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.gene_selection.load_gtf">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">load_gtf</span><span class="p">(</span><span class="n">gtf_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load gene coordinates from a GTF file into a pandas DataFrame using gtfparse.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gtf_path : pathlib.Path</span>
<span class="sd">        The path to the GTF file (optionally gzipped).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        A DataFrame containing parsed GTF records. The columns correspond to</span>
<span class="sd">        GTF fields such as &quot;gene_name&quot;, &quot;gene_type&quot;, &quot;start&quot;, &quot;end&quot;, etc.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading GTF from </span><span class="si">{</span><span class="n">gtf_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">read_gtf</span><span class="p">(</span><span class="n">gtf_path</span><span class="p">)</span>
    <span class="n">gene_coordinates</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">gene_coordinates</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span>
    <span class="k">return</span> <span class="n">gene_coordinates</span></div>


<div class="viewcode-block" id="filter_protein_coding_genes">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.gene_selection.filter_protein_coding_genes">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">filter_protein_coding_genes</span><span class="p">(</span>
    <span class="n">data_rna</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">gtf_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retain only protein-coding genes in the RNA AnnData object based on GTF annotations.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_rna : anndata.AnnData</span>
<span class="sd">        RNA single-cell data.</span>
<span class="sd">    gtf_df : pd.DataFrame</span>
<span class="sd">        A GTF DataFrame (from `load_gtf`) containing columns like &quot;gene_type&quot;, &quot;gene_name&quot;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    anndata.AnnData</span>
<span class="sd">        The AnnData subset containing only protein-coding genes found in both</span>
<span class="sd">        the original data and the GTF&#39;s &quot;gene_type == &#39;protein_coding&#39;&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df_protein_coding</span> <span class="o">=</span> <span class="n">gtf_df</span><span class="p">[</span><span class="n">gtf_df</span><span class="p">[</span><span class="s2">&quot;gene_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;protein_coding&quot;</span><span class="p">]</span>
    <span class="n">pc_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">df_protein_coding</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
    <span class="n">rna_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="n">keep_genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">pc_genes</span> <span class="o">&amp;</span> <span class="n">rna_genes</span><span class="p">))</span>
    <span class="n">data_rna_sub</span> <span class="o">=</span> <span class="n">data_rna</span><span class="p">[:,</span> <span class="n">keep_genes</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtered to protein-coding genes: </span><span class="si">{</span><span class="n">data_rna_sub</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2"> genes left.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data_rna_sub</span></div>


<div class="viewcode-block" id="compute_hvgs_and_tfs">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.gene_selection.compute_hvgs_and_tfs">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_hvgs_and_tfs</span><span class="p">(</span>
    <span class="n">data_rna</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">tf_names</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">user_genes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">user_tfs</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">num_genes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3000</span><span class="p">,</span>
    <span class="n">num_tfs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
    <span class="n">min_cells</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute sets of Highly Variable Genes (HVGs) and TFs (transcription factors)</span>
<span class="sd">    for scDoRI training.</span>

<span class="sd">    This function:</span>
<span class="sd">      1. Identifies user-specified genes and TFs present in `data_rna`.</span>
<span class="sd">      2. Selects additional TFs by computing HVGs among potential TFs up to `num_tfs`.</span>
<span class="sd">      3. Selects non-TF HVGs up to `num_genes` (minus any user-specified genes and TFs).</span>
<span class="sd">      4. Combines these sets into a final AnnData subset and returns them.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_rna : anndata.AnnData</span>
<span class="sd">        The RNA single-cell data from which to select HVGs and TFs.</span>
<span class="sd">    tf_names : list of str</span>
<span class="sd">        A list of all TF names (from a motif database or known TF list).</span>
<span class="sd">    user_genes : list of str, optional</span>
<span class="sd">        A list of user-specified genes that must be included in the final set,</span>
<span class="sd">        default is None.</span>
<span class="sd">    user_tfs : list of str, optional</span>
<span class="sd">        A list of user-specified TFs that must be included in the final set,</span>
<span class="sd">        default is None.</span>
<span class="sd">    num_genes : int, optional</span>
<span class="sd">        Total number of HVGs (non-TFs) desired. Default is 3000.</span>
<span class="sd">    num_tfs : int, optional</span>
<span class="sd">        Total number of TFs desired. Default is 300.</span>
<span class="sd">    min_cells : int, optional</span>
<span class="sd">        Minimum number of cells in which a gene must be detected (e.g., nonzero)</span>
<span class="sd">        to be considered for HVG selection. Default is 20 (not fully enforced in</span>
<span class="sd">        this code snippet, but typically used with standard HVG filtering).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    data_rna_processed : anndata.AnnData</span>
<span class="sd">        A subset of the original data containing the selected HVGs and TFs.</span>
<span class="sd">    final_genes : list of str</span>
<span class="sd">        The final list of HVGs (non-TFs).</span>
<span class="sd">    final_tfs : list of str</span>
<span class="sd">        The final list of TFs.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - HVG selection is done by `scanpy.pp.highly_variable_genes`, using normalized/log1p data.</span>
<span class="sd">    - User-provided genes and TFs are included by default, removing them from the</span>
<span class="sd">      HVG candidate pool if they were already selected.</span>
<span class="sd">    - TFs are not re-labeled or otherwise changed beyond this classification. </span>
<span class="sd">    - The column `data_rna_processed.var[&quot;gene_type&quot;]` is set to &quot;HVG&quot; or &quot;TF&quot; for each gene.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">user_genes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">user_genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">user_tfs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">user_tfs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Selecting HVGs and TFs...&quot;</span><span class="p">)</span>

    <span class="c1"># 1) Validate user-specified lists</span>
    <span class="n">valid_genes_user</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">user_genes</span><span class="p">))</span>
    <span class="n">valid_tfs_user</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
        <span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">user_tfs</span><span class="p">)</span>
        <span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">tf_names</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">num_tfs_hvg</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">num_tfs</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">valid_tfs_user</span><span class="p">))</span>
    <span class="n">num_genes_hvg</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">num_genes</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">valid_genes_user</span><span class="p">)</span> <span class="o">-</span> <span class="n">num_tfs</span><span class="p">)</span>

    <span class="c1"># 2) HVGs among TFs</span>
    <span class="n">tf_candidates</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">((</span><span class="nb">set</span><span class="p">(</span><span class="n">tf_names</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">valid_tfs_user</span><span class="p">))</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)))</span>
    <span class="n">data_rna_tf</span> <span class="o">=</span> <span class="n">data_rna</span><span class="p">[:,</span> <span class="n">tf_candidates</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">data_rna_tf</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">data_rna_tf</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">data_rna_tf</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="n">num_tfs_hvg</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">selected_tfs</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">data_rna_tf</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span> <span class="o">+</span> <span class="n">valid_tfs_user</span><span class="p">)</span>

    <span class="c1"># 3) HVGs among non-TFs</span>
    <span class="n">non_tf_candidates</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">selected_tfs</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">valid_genes_user</span><span class="p">)</span>
    <span class="n">data_rna_non_tf</span> <span class="o">=</span> <span class="n">data_rna</span><span class="p">[:,</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">non_tf_candidates</span><span class="p">))]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">data_rna_non_tf</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">data_rna_non_tf</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">data_rna_non_tf</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="n">num_genes_hvg</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">selected_non_tfs</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">data_rna_non_tf</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">valid_genes_user</span><span class="p">))</span>
    <span class="n">selected_non_tfs</span> <span class="o">=</span> <span class="p">[</span><span class="n">g</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">selected_non_tfs</span> <span class="k">if</span> <span class="n">g</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">selected_tfs</span><span class="p">]</span>

    <span class="n">final_genes</span> <span class="o">=</span> <span class="n">selected_non_tfs</span>
    <span class="n">final_tfs</span> <span class="o">=</span> <span class="n">selected_tfs</span>

    <span class="n">combined</span> <span class="o">=</span> <span class="n">final_genes</span> <span class="o">+</span> <span class="n">final_tfs</span>
    <span class="n">data_rna_processed</span> <span class="o">=</span> <span class="n">data_rna</span><span class="p">[:,</span> <span class="n">combined</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Mark gene_type in .var</span>
    <span class="n">gene_types</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;HVG&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">final_genes</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;TF&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">final_tfs</span><span class="p">)</span>
    <span class="n">data_rna_processed</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;gene_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_types</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Selected </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">final_genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> HVGs + </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">final_tfs</span><span class="p">)</span><span class="si">}</span><span class="s2"> TFs.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data_rna_processed</span><span class="p">,</span> <span class="n">final_genes</span><span class="p">,</span> <span class="n">final_tfs</span></div>

</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">scDoRI</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../preprocessing_pipeline.html">Preprocessing Pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../scdori.html">scDORI training</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../notebooks.html">Notebook Examples</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2025, Manu Saraswat.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
    </div>

    

    
  </body>
</html>