
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>preprocessing_pipeline.utils &#8212; scDoRI  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/preprocessing_pipeline/utils';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">scDoRI  documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../api_reference.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar hide-on-wide">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../api_reference.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">Module code</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">preprocessing_pipeline.utils</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for preprocessing_pipeline.utils</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">itertools</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<div class="viewcode-block" id="create_extended_gene_bed">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.utils.create_extended_gene_bed">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">create_extended_gene_bed</span><span class="p">(</span>
    <span class="n">gtf_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">final_genes</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">chrom_sizes_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create an extended gene BED-like DataFrame by adding a window around each gene.</span>

<span class="sd">    For each gene in `final_genes`, take the gene feature from `gtf_df` and extend</span>
<span class="sd">    the start and end coordinates by `window_size`. If a chromosome sizes file is</span>
<span class="sd">    provided, clamp the extended coordinates within valid chromosome boundaries.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gtf_df : pd.DataFrame</span>
<span class="sd">        A DataFrame containing GTF entries (e.g., loaded by gtfparse). Expected columns:</span>
<span class="sd">        [&quot;feature&quot;, &quot;gene_name&quot;, &quot;seqname&quot;, &quot;start&quot;, &quot;end&quot;, &quot;strand&quot;, ...].</span>
<span class="sd">    final_genes : list of str</span>
<span class="sd">        The list of gene names to be extended.</span>
<span class="sd">    window_size : int</span>
<span class="sd">        The number of base pairs to extend on each side of the gene.</span>
<span class="sd">    chrom_sizes_path : pathlib.Path, optional</span>
<span class="sd">        A path to a tab-delimited file of chromosome sizes (e.g., from UCSC),</span>
<span class="sd">        with columns [chrom, size]. If provided, coordinates are clamped to [0, chrom_size].</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    pd.DataFrame</span>
<span class="sd">        A DataFrame with columns [&quot;chr&quot;, &quot;start_new&quot;, &quot;end_new&quot;, &quot;gene_name&quot;, &quot;strand&quot;]</span>
<span class="sd">        representing the extended regions for each gene.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Only rows with `feature == &quot;gene&quot;` are considered in the GTF.</span>
<span class="sd">    - If a gene is not found in `gtf_df`, it is omitted.</span>
<span class="sd">    - The `chr` column is set from the GTF&#39;s &quot;seqname&quot; field.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter for final genes &amp; gene features</span>
    <span class="n">gtf_gene</span> <span class="o">=</span> <span class="n">gtf_df</span><span class="p">[</span><span class="n">gtf_df</span><span class="p">[</span><span class="s2">&quot;feature&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s2">&quot;gene_name&quot;</span><span class="p">)</span>
    <span class="n">gtf_gene</span> <span class="o">=</span> <span class="n">gtf_gene</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;gene_name&quot;</span><span class="p">)</span>
    <span class="n">gtf_gene</span> <span class="o">=</span> <span class="n">gtf_gene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">final_genes</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">gtf_gene</span><span class="o">.</span><span class="n">index</span><span class="p">))]</span>

    <span class="c1"># rename for clarity</span>
    <span class="n">gtf_gene</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gtf_gene</span><span class="p">[</span><span class="s2">&quot;seqname&quot;</span><span class="p">]</span>

    <span class="c1"># If we want clamping, load chrom.sizes</span>
    <span class="n">chrom_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">chrom_sizes_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">chrom_sizes_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">chrom_sizes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">chrom_sizes_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;size&quot;</span><span class="p">])</span>
        <span class="n">chrom_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">chrom_sizes</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">chrom_sizes</span><span class="p">[</span><span class="s2">&quot;size&quot;</span><span class="p">]))</span>

    <span class="n">extended_rows</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gtf_gene</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">gtf_gene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span>
        <span class="n">chr_</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span>
        <span class="n">start_</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
        <span class="n">end_</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
        <span class="n">strand_</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="s2">&quot;strand&quot;</span> <span class="ow">in</span> <span class="n">row</span> <span class="k">else</span> <span class="s2">&quot;+&quot;</span>

        <span class="n">start_extended</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">start_</span> <span class="o">-</span> <span class="n">window_size</span><span class="p">)</span>
        <span class="n">end_extended</span> <span class="o">=</span> <span class="n">end_</span> <span class="o">+</span> <span class="n">window_size</span>

        <span class="k">if</span> <span class="n">chr_</span> <span class="ow">in</span> <span class="n">chrom_dict</span><span class="p">:</span>
            <span class="n">chr_len</span> <span class="o">=</span> <span class="n">chrom_dict</span><span class="p">[</span><span class="n">chr_</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">end_extended</span> <span class="o">&gt;</span> <span class="n">chr_len</span><span class="p">:</span>
                <span class="n">end_extended</span> <span class="o">=</span> <span class="n">chr_len</span>
            <span class="c1"># start_extended already clamped to 0 above</span>

        <span class="n">extended_rows</span><span class="o">.</span><span class="n">append</span><span class="p">([</span>
            <span class="n">chr_</span><span class="p">,</span>
            <span class="n">start_extended</span><span class="p">,</span>
            <span class="n">end_extended</span><span class="p">,</span>
            <span class="n">gene</span><span class="p">,</span>
            <span class="n">strand_</span>
        <span class="p">])</span>

    <span class="n">df_extended</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">extended_rows</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;start_new&quot;</span><span class="p">,</span> <span class="s2">&quot;end_new&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_name&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">df_extended</span></div>


<div class="viewcode-block" id="compute_gene_peak_distance_matrix">
<a class="viewcode-back" href="../../preprocessing_pipeline.html#preprocessing_pipeline.utils.compute_gene_peak_distance_matrix">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compute_gene_peak_distance_matrix</span><span class="p">(</span>
    <span class="n">data_atac</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">data_rna</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">gene_coordinates_intersect</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute a distance matrix between peaks (ATAC) and genes (RNA), accounting for strand.</span>

<span class="sd">    This function:</span>
<span class="sd">    -----</span>
<span class="sd">    1. Expects the ATAC data (`data_atac.var`) to have columns &quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, and &quot;peak_name&quot;.</span>
<span class="sd">    2. Expects the gene coordinates DataFrame (`gene_coordinates_intersect`) to have columns</span>
<span class="sd">        [&quot;chr_gene&quot;, &quot;start&quot;, &quot;end&quot;, &quot;strand&quot;, &quot;gene&quot;].</span>
<span class="sd">    3. For each gene, we compute a distance to each peak by:</span>
<span class="sd">        - 0 if the peak midpoint is within gene-body or promoter (5kb upstream of TSS by default),</span>
<span class="sd">        e.g., if on the &quot;+&quot; strand and midpoint is between (start - 5000) and end =&gt; distance=0.</span>
<span class="sd">        If on the &quot;-&quot; strand and midpoint is between start and (end + 5000) =&gt; distance=0.</span>
<span class="sd">        - Otherwise, distance = min(|mid - start|, |mid - end|).</span>
<span class="sd">        - If the chromosome of the gene != chromosome of the peak =&gt; distance = -1.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data_atac : pd.DataFrame</span>
<span class="sd">        The ATAC dataset. Its `.var` should contain &quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;peak_name&quot;.</span>
<span class="sd">    data_rna : pd.DataFrame</span>
<span class="sd">        The RNA dataset. Its `.var` is expected to have gene names as indices or relevant columns.</span>
<span class="sd">    gene_coordinates_intersect : pd.DataFrame</span>
<span class="sd">        A DataFrame with gene info, columns at least:</span>
<span class="sd">        [&quot;chr_gene&quot;, &quot;start&quot;, &quot;end&quot;, &quot;strand&quot;, &quot;gene&quot;].</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    np.ndarray</span>
<span class="sd">        A 2D NumPy array of shape (n_genes, n_peaks) representing gene-peak distances.</span>
<span class="sd">        Each row corresponds to a gene (matching `data_rna.var.index`),</span>
<span class="sd">        and each column corresponds to a peak (matching `data_atac.var.index` order).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Starting computation of gene-peak distances...&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of genes: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="si">}</span><span class="s2">, Number of peaks: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ATAC var columns: </span><span class="si">{</span><span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="s1">&#39;chr&#39;</span> <span class="ow">in</span> <span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;chr&#39; column is missing in data_atac.var!&quot;</span>
    <span class="k">assert</span> <span class="s1">&#39;start&#39;</span> <span class="ow">in</span> <span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;start&#39; column is missing in data_atac.var!&quot;</span>
    <span class="k">assert</span> <span class="s1">&#39;end&#39;</span> <span class="ow">in</span> <span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;end&#39; column is missing in data_atac.var!&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene coordinates columns: </span><span class="si">{</span><span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">columns</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="s1">&#39;chr_gene&#39;</span> <span class="ow">in</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;chr_gene&#39; column is missing!&quot;</span>
    <span class="k">assert</span> <span class="s1">&#39;start&#39;</span> <span class="ow">in</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;start&#39; column is missing!&quot;</span>
    <span class="k">assert</span> <span class="s1">&#39;end&#39;</span> <span class="ow">in</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;end&#39; column is missing!&quot;</span>
    <span class="k">assert</span> <span class="s1">&#39;strand&#39;</span> <span class="ow">in</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="s2">&quot;&#39;strand&#39; column is missing!&quot;</span>

    <span class="n">peak_gene_distance</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Add midpoints to peaks in the ATAC data</span>
    <span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;mid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">])</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Added midpoint column to ATAC peaks.&quot;</span><span class="p">)</span>

    <span class="c1"># Iterate over all genes</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">data_rna</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)):</span>
        <span class="n">chr_gene</span> <span class="o">=</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;chr_gene&quot;</span><span class="p">]</span>
        <span class="n">start_gene</span> <span class="o">=</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">]</span>
        <span class="n">end_gene</span> <span class="o">=</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]</span>
        <span class="n">strand_gene</span> <span class="o">=</span> <span class="n">gene_coordinates_intersect</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">]</span>

        <span class="c1"># Build pairs of all peaks with this gene</span>
        <span class="n">atac_var_all</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;peak_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="p">[</span><span class="n">gene</span><span class="p">]),</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;peak_name&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">]</span>
        <span class="p">)</span>

        <span class="c1"># Merge to get peak coordinates</span>
        <span class="n">atac_var_all</span> <span class="o">=</span> <span class="n">atac_var_all</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">data_atac</span><span class="o">.</span><span class="n">var</span><span class="p">[[</span><span class="s2">&quot;peak_name&quot;</span><span class="p">,</span> <span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;mid&quot;</span><span class="p">]],</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;peak_name&quot;</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span>
        <span class="p">)</span>
        <span class="n">atac_var_all</span> <span class="o">=</span> <span class="n">atac_var_all</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
            <span class="n">gene_coordinates_intersect</span><span class="p">[[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_gene&quot;</span><span class="p">,</span> <span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">]],</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span>
        <span class="p">)</span>

        <span class="n">mid</span> <span class="o">=</span> <span class="n">atac_var_all</span><span class="p">[</span><span class="s2">&quot;mid&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">start_arr</span> <span class="o">=</span> <span class="n">atac_var_all</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">end_arr</span> <span class="o">=</span> <span class="n">atac_var_all</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="k">if</span> <span class="n">strand_gene</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
            <span class="c1"># Zero distance if peak midpoint is between [start_gene - 5000, end_gene]</span>
            <span class="n">within_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">mid</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">start_arr</span> <span class="o">-</span> <span class="mi">5000</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">mid</span> <span class="o">&lt;=</span> <span class="n">end_arr</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Zero distance if peak midpoint is between [start_gene, end_gene + 5000]</span>
            <span class="n">within_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">mid</span> <span class="o">&gt;=</span> <span class="n">start_arr</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">mid</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="n">end_arr</span> <span class="o">+</span> <span class="mi">5000</span><span class="p">))</span>

        <span class="n">dist_to_start</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">mid</span> <span class="o">-</span> <span class="n">start_arr</span><span class="p">)</span>
        <span class="n">dist_to_end</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">mid</span> <span class="o">-</span> <span class="n">end_arr</span><span class="p">)</span>
        <span class="n">min_dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">dist_to_start</span><span class="p">,</span> <span class="n">dist_to_end</span><span class="p">)</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">within_mask</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">min_dist</span><span class="p">)</span>

        <span class="c1"># If chromosome doesn&#39;t match</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span>
            <span class="n">atac_var_all</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">==</span> <span class="n">atac_var_all</span><span class="p">[</span><span class="s2">&quot;chr_gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="n">distance</span><span class="p">,</span>
            <span class="o">-</span><span class="mi">1</span>
        <span class="p">)</span>

        <span class="n">peak_gene_distance</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span>

        <span class="c1"># Optional debug checkpoint</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">peak_gene_distance</span><span class="p">)</span> <span class="o">%</span> <span class="mi">100</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">peak_gene_distance</span><span class="p">)</span><span class="si">}</span><span class="s2"> genes so far.&quot;</span><span class="p">)</span>

    <span class="c1"># Convert to NumPy array =&gt; shape (n_genes, n_peaks)</span>
    <span class="n">peak_gene_distance_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">peak_gene_distance</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Gene-peak distance matrix computed with shape: </span><span class="si">{</span><span class="n">peak_gene_distance_matrix</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">peak_gene_distance_matrix</span></div>

</pre></div>

                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Manu Saraswat.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>